// @react-grab/cli@0.1.31/dist/cli.cjs
#!/usr/bin/env node
//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));
//#endregion
let commander = require("commander");
let picocolors = require("picocolors");
picocolors = __toESM(picocolors);
let node_fs = require("node:fs");
node_fs = __toESM(node_fs);
let node_path = require("node:path");
node_path = __toESM(node_path);
let _antfu_ni = require("@antfu/ni");
let ignore = require("ignore");
ignore = __toESM(ignore);
let node_os = require("node:os");
node_os = __toESM(node_os);
let node_process = require("node:process");
node_process = __toESM(node_process);
let jsonc_parser = require("jsonc-parser");
jsonc_parser = __toESM(jsonc_parser);
let smol_toml = require("smol-toml");
smol_toml = __toESM(smol_toml);
let prompts = require("prompts");
prompts = __toESM(prompts);
let ora = require("ora");
ora = __toESM(ora);
let node_child_process = require("node:child_process");
//#region src/utils/is-non-interactive.ts
const AGENT_ENVIRONMENT_VARIABLES = [
	"CI",
	"CLAUDECODE",
	"CURSOR_AGENT",
	"CODEX_CI",
	"OPENCODE",
	"AMP_HOME",
	"AMI"
];
const isEnvironmentVariableSet = (variable) => Boolean(process.env[variable]);
const detectNonInteractive = (yesFlag) => yesFlag || AGENT_ENVIRONMENT_VARIABLES.some(isEnvironmentVariableSet) || !process.stdin.isTTY;
//#endregion
//#region src/utils/detect.ts
const VALID_PACKAGE_MANAGERS = new Set([
	"npm",
	"yarn",
	"pnpm",
	"bun"
]);
const detectPackageManager = async (projectRoot) => {
	const detected = await (0, _antfu_ni.detect)({ cwd: projectRoot });
	if (detected) {
		const managerName = detected.split("@")[0];
		if (VALID_PACKAGE_MANAGERS.has(managerName)) return managerName;
	}
	return "npm";
};
const detectFramework = (projectRoot) => {
	const packageJsonPath = (0, node_path.join)(projectRoot, "package.json");
	if (!(0, node_fs.existsSync)(packageJsonPath)) return "unknown";
	try {
		const packageJson = JSON.parse((0, node_fs.readFileSync)(packageJsonPath, "utf-8"));
		const allDependencies = {
			...packageJson.dependencies,
			...packageJson.devDependencies
		};
		if (allDependencies["next"]) return "next";
		if (allDependencies["@tanstack/react-start"]) return "tanstack";
		if (allDependencies["vite"]) return "vite";
		if (allDependencies["webpack"]) return "webpack";
		return "unknown";
	} catch {
		return "unknown";
	}
};
const detectNextRouterType = (projectRoot) => {
	const hasAppDir = (0, node_fs.existsSync)((0, node_path.join)(projectRoot, "app"));
	const hasSrcAppDir = (0, node_fs.existsSync)((0, node_path.join)(projectRoot, "src", "app"));
	const hasPagesDir = (0, node_fs.existsSync)((0, node_path.join)(projectRoot, "pages"));
	const hasSrcPagesDir = (0, node_fs.existsSync)((0, node_path.join)(projectRoot, "src", "pages"));
	if (hasAppDir || hasSrcAppDir) return "app";
	if (hasPagesDir || hasSrcPagesDir) return "pages";
	return "unknown";
};
const detectMonorepo = (projectRoot) => {
	if ((0, node_fs.existsSync)((0, node_path.join)(projectRoot, "pnpm-workspace.yaml"))) return true;
	if ((0, node_fs.existsSync)((0, node_path.join)(projectRoot, "lerna.json"))) return true;
	const packageJsonPath = (0, node_path.join)(projectRoot, "package.json");
	if ((0, node_fs.existsSync)(packageJsonPath)) try {
		if (JSON.parse((0, node_fs.readFileSync)(packageJsonPath, "utf-8")).workspaces) return true;
	} catch {
		return false;
	}
	return false;
};
const getWorkspacePatterns = (projectRoot) => {
	const patterns = [];
	const pnpmWorkspacePath = (0, node_path.join)(projectRoot, "pnpm-workspace.yaml");
	if ((0, node_fs.existsSync)(pnpmWorkspacePath)) {
		const lines = (0, node_fs.readFileSync)(pnpmWorkspacePath, "utf-8").split("\n");
		let inPackages = false;
		for (const line of lines) {
			if (line.match(/^packages:\s*$/)) {
				inPackages = true;
				continue;
			}
			if (inPackages) {
				if (line.match(/^[a-zA-Z]/) || line.trim() === "") {
					if (line.match(/^[a-zA-Z]/)) inPackages = false;
					continue;
				}
				const match = line.match(/^\s*-\s*['"]?([^'"#\n]+?)['"]?\s*$/);
				if (match) patterns.push(match[1].trim());
			}
		}
	}
	const lernaJsonPath = (0, node_path.join)(projectRoot, "lerna.json");
	if ((0, node_fs.existsSync)(lernaJsonPath)) try {
		const lernaJson = JSON.parse((0, node_fs.readFileSync)(lernaJsonPath, "utf-8"));
		if (Array.isArray(lernaJson.packages)) patterns.push(...lernaJson.packages);
	} catch {}
	const packageJsonPath = (0, node_path.join)(projectRoot, "package.json");
	if ((0, node_fs.existsSync)(packageJsonPath)) try {
		const packageJson = JSON.parse((0, node_fs.readFileSync)(packageJsonPath, "utf-8"));
		if (Array.isArray(packageJson.workspaces)) patterns.push(...packageJson.workspaces);
		else if (packageJson.workspaces?.packages) patterns.push(...packageJson.workspaces.packages);
	} catch {}
	return [...new Set(patterns)];
};
const expandWorkspacePattern = (projectRoot, pattern) => {
	const isGlob = pattern.endsWith("/*");
	const basePath = (0, node_path.join)(projectRoot, pattern.replace(/\/\*$/, ""));
	if (!(0, node_fs.existsSync)(basePath)) return [];
	if (!isGlob) return (0, node_fs.existsSync)((0, node_path.join)(basePath, "package.json")) ? [basePath] : [];
	const results = [];
	try {
		const entries = (0, node_fs.readdirSync)(basePath, { withFileTypes: true });
		for (const entry of entries) {
			if (!entry.isDirectory()) continue;
			if ((0, node_fs.existsSync)((0, node_path.join)(basePath, entry.name, "package.json"))) results.push((0, node_path.join)(basePath, entry.name));
		}
	} catch {
		return results;
	}
	return results;
};
const hasReactDependency = (projectPath) => {
	const packageJsonPath = (0, node_path.join)(projectPath, "package.json");
	if (!(0, node_fs.existsSync)(packageJsonPath)) return false;
	try {
		const packageJson = JSON.parse((0, node_fs.readFileSync)(packageJsonPath, "utf-8"));
		const allDeps = {
			...packageJson.dependencies,
			...packageJson.devDependencies
		};
		return Boolean(allDeps["react"] || allDeps["react-dom"]);
	} catch {
		return false;
	}
};
const buildReactProject = (projectPath) => {
	const framework = detectFramework(projectPath);
	if (!hasReactDependency(projectPath) && framework === "unknown") return null;
	let name = (0, node_path.basename)(projectPath);
	const packageJsonPath = (0, node_path.join)(projectPath, "package.json");
	try {
		name = JSON.parse((0, node_fs.readFileSync)(packageJsonPath, "utf-8")).name || name;
	} catch {}
	return {
		name,
		path: projectPath,
		framework
	};
};
const findWorkspaceProjects = (projectRoot) => {
	const patterns = getWorkspacePatterns(projectRoot);
	const projects = [];
	for (const pattern of patterns) for (const projectPath of expandWorkspacePattern(projectRoot, pattern)) {
		const project = buildReactProject(projectPath);
		if (project) projects.push(project);
	}
	return projects;
};
const ALWAYS_IGNORED_DIRECTORIES = [
	"node_modules",
	".git",
	".next",
	".cache",
	".turbo",
	"dist",
	"build",
	"coverage",
	"test-results"
];
const loadGitignore = (projectRoot) => {
	const ignorer = (0, ignore.default)().add(ALWAYS_IGNORED_DIRECTORIES);
	const gitignorePath = (0, node_path.join)(projectRoot, ".gitignore");
	if ((0, node_fs.existsSync)(gitignorePath)) try {
		ignorer.add((0, node_fs.readFileSync)(gitignorePath, "utf-8"));
	} catch {}
	return ignorer;
};
const scanDirectoryForProjects = (rootDirectory, ignorer, maxDepth, currentDepth = 0) => {
	if (currentDepth >= maxDepth) return [];
	if (!(0, node_fs.existsSync)(rootDirectory)) return [];
	const projects = [];
	try {
		const entries = (0, node_fs.readdirSync)(rootDirectory, { withFileTypes: true });
		for (const entry of entries) {
			if (!entry.isDirectory()) continue;
			if (ignorer.ignores(entry.name)) continue;
			const entryPath = (0, node_path.join)(rootDirectory, entry.name);
			if ((0, node_fs.existsSync)((0, node_path.join)(entryPath, "package.json"))) {
				const project = buildReactProject(entryPath);
				if (project) {
					projects.push(project);
					continue;
				}
			}
			projects.push(...scanDirectoryForProjects(entryPath, ignorer, maxDepth, currentDepth + 1));
		}
	} catch {
		return projects;
	}
	return projects;
};
const MAX_SCAN_DEPTH = 2;
const findReactProjects = (projectRoot) => {
	if (detectMonorepo(projectRoot)) {
		const workspaceProjects = findWorkspaceProjects(projectRoot);
		if (workspaceProjects.length > 0) return workspaceProjects;
	}
	const scannedProjects = scanDirectoryForProjects(projectRoot, loadGitignore(projectRoot), MAX_SCAN_DEPTH);
	if (scannedProjects.length > 0) return scannedProjects;
	let currentDirectory = (0, node_path.dirname)(projectRoot);
	while (currentDirectory !== (0, node_path.dirname)(currentDirectory)) {
		const parentProject = buildReactProject(currentDirectory);
		if (parentProject) return [parentProject];
		currentDirectory = (0, node_path.dirname)(currentDirectory);
	}
	return [];
};
const hasReactGrabInFile = (filePath) => {
	if (!(0, node_fs.existsSync)(filePath)) return false;
	try {
		const content = (0, node_fs.readFileSync)(filePath, "utf-8");
		return [
			/["'`][^"'`]*react-grab/,
			/react-grab[^"'`]*["'`]/,
			/<[^>]*react-grab/i,
			/import[^;]*react-grab/i,
			/require[^)]*react-grab/i,
			/from\s+[^;]*react-grab/i,
			/src[^>]*react-grab/i
		].some((pattern) => pattern.test(content));
	} catch {
		return false;
	}
};
const detectReactGrab = (projectRoot) => {
	const packageJsonPath = (0, node_path.join)(projectRoot, "package.json");
	if ((0, node_fs.existsSync)(packageJsonPath)) try {
		const packageJson = JSON.parse((0, node_fs.readFileSync)(packageJsonPath, "utf-8"));
		if ({
			...packageJson.dependencies,
			...packageJson.devDependencies
		}["react-grab"]) return true;
	} catch {}
	return [
		(0, node_path.join)(projectRoot, "app", "layout.tsx"),
		(0, node_path.join)(projectRoot, "app", "layout.jsx"),
		(0, node_path.join)(projectRoot, "src", "app", "layout.tsx"),
		(0, node_path.join)(projectRoot, "src", "app", "layout.jsx"),
		(0, node_path.join)(projectRoot, "pages", "_document.tsx"),
		(0, node_path.join)(projectRoot, "pages", "_document.jsx"),
		(0, node_path.join)(projectRoot, "instrumentation-client.ts"),
		(0, node_path.join)(projectRoot, "instrumentation-client.js"),
		(0, node_path.join)(projectRoot, "src", "instrumentation-client.ts"),
		(0, node_path.join)(projectRoot, "src", "instrumentation-client.js"),
		(0, node_path.join)(projectRoot, "index.html"),
		(0, node_path.join)(projectRoot, "public", "index.html"),
		(0, node_path.join)(projectRoot, "src", "index.tsx"),
		(0, node_path.join)(projectRoot, "src", "index.ts"),
		(0, node_path.join)(projectRoot, "src", "main.tsx"),
		(0, node_path.join)(projectRoot, "src", "main.ts"),
		(0, node_path.join)(projectRoot, "src", "routes", "__root.tsx"),
		(0, node_path.join)(projectRoot, "src", "routes", "__root.jsx"),
		(0, node_path.join)(projectRoot, "app", "routes", "__root.tsx"),
		(0, node_path.join)(projectRoot, "app", "routes", "__root.jsx")
	].some(hasReactGrabInFile);
};
const detectUnsupportedFramework = (projectRoot) => {
	const packageJsonPath = (0, node_path.join)(projectRoot, "package.json");
	if (!(0, node_fs.existsSync)(packageJsonPath)) return null;
	try {
		const packageJson = JSON.parse((0, node_fs.readFileSync)(packageJsonPath, "utf-8"));
		const allDependencies = {
			...packageJson.dependencies,
			...packageJson.devDependencies
		};
		if (allDependencies["@remix-run/react"] || allDependencies["remix"]) return "remix";
		if (allDependencies["astro"]) return "astro";
		if (allDependencies["@sveltejs/kit"]) return "sveltekit";
		if (allDependencies["gatsby"]) return "gatsby";
		return null;
	} catch {
		return null;
	}
};
const detectReactGrabVersion = (projectRoot) => {
	const installedPackageJsonPath = (0, node_path.join)(projectRoot, "node_modules", "react-grab", "package.json");
	if ((0, node_fs.existsSync)(installedPackageJsonPath)) try {
		return JSON.parse((0, node_fs.readFileSync)(installedPackageJsonPath, "utf-8")).version ?? null;
	} catch {}
	return null;
};
const detectProject = async (projectRoot = process.cwd()) => {
	const framework = detectFramework(projectRoot);
	return {
		packageManager: await detectPackageManager(projectRoot),
		framework,
		nextRouterType: framework === "next" ? detectNextRouterType(projectRoot) : "unknown",
		isMonorepo: detectMonorepo(projectRoot),
		projectRoot,
		hasReactGrab: detectReactGrab(projectRoot),
		reactGrabVersion: detectReactGrabVersion(projectRoot),
		unsupportedFramework: detectUnsupportedFramework(projectRoot)
	};
};
//#endregion
//#region src/utils/highlighter.ts
const highlighter = {
	error: picocolors.default.red,
	warn: picocolors.default.yellow,
	info: picocolors.default.cyan,
	success: picocolors.default.green,
	dim: picocolors.default.dim
};
//#endregion
//#region src/utils/logger.ts
const logger = {
	error(...args) {
		console.log(highlighter.error(args.join(" ")));
	},
	warn(...args) {
		console.log(highlighter.warn(args.join(" ")));
	},
	success(...args) {
		console.log(highlighter.success(args.join(" ")));
	},
	log(...args) {
		console.log(args.join(" "));
	},
	break() {
		console.log("");
	}
};
//#endregion
//#region src/utils/handle-error.ts
const handleError = (error) => {
	logger.break();
	logger.error("Something went wrong. Please check the error below for more details.");
	logger.error("If the problem persists, please open an issue on GitHub.");
	logger.error("");
	if (error instanceof Error) logger.error(error.message);
	logger.break();
	process.exit(1);
};
//#endregion
//#region src/utils/prompts.ts
const onCancel = () => {
	logger.break();
	logger.log("Cancelled.");
	logger.break();
	process.exit(0);
};
const prompts$1 = (questions) => {
	return (0, prompts.default)(questions, { onCancel });
};
//#endregion
//#region src/utils/spinner.ts
const spinner = (text) => (0, ora.default)({ text });
//#endregion
//#region src/utils/install-mcp.ts
const SERVER_NAME = "react-grab-mcp";
const PACKAGE_NAME = "@react-grab/mcp";
const getXdgConfigHome = () => node_process.default.env.XDG_CONFIG_HOME || node_path.default.join(node_os.default.homedir(), ".config");
const getBaseDir = () => {
	const homeDir = node_os.default.homedir();
	if (node_process.default.platform === "win32") return node_process.default.env.APPDATA || node_path.default.join(homeDir, "AppData", "Roaming");
	if (node_process.default.platform === "darwin") return node_path.default.join(homeDir, "Library", "Application Support");
	return getXdgConfigHome();
};
const getZedConfigPath = () => {
	if (node_process.default.platform === "win32") return node_path.default.join(getBaseDir(), "Zed", "settings.json");
	return node_path.default.join(node_os.default.homedir(), ".config", "zed", "settings.json");
};
const getOpenCodeConfigPath = () => {
	const configDir = node_path.default.join(getXdgConfigHome(), "opencode");
	const jsoncPath = node_path.default.join(configDir, "opencode.jsonc");
	const jsonPath = node_path.default.join(configDir, "opencode.json");
	if (node_fs.default.existsSync(jsoncPath)) return jsoncPath;
	if (node_fs.default.existsSync(jsonPath)) return jsonPath;
	return jsoncPath;
};
const getClients = () => {
	const homeDir = node_os.default.homedir();
	const baseDir = getBaseDir();
	const stdioConfig = {
		command: "npx",
		args: [
			"-y",
			PACKAGE_NAME,
			"--stdio"
		]
	};
	return [
		{
			name: "Claude Code",
			configPath: node_path.default.join(homeDir, ".claude.json"),
			configKey: "mcpServers",
			format: "json",
			serverConfig: stdioConfig
		},
		{
			name: "Codex",
			configPath: node_path.default.join(node_process.default.env.CODEX_HOME || node_path.default.join(homeDir, ".codex"), "config.toml"),
			configKey: "mcp_servers",
			format: "toml",
			serverConfig: stdioConfig
		},
		{
			name: "Cursor",
			configPath: node_path.default.join(homeDir, ".cursor", "mcp.json"),
			configKey: "mcpServers",
			format: "json",
			serverConfig: stdioConfig
		},
		{
			name: "OpenCode",
			configPath: getOpenCodeConfigPath(),
			configKey: "mcp",
			format: "json",
			serverConfig: {
				type: "local",
				command: [
					"npx",
					"-y",
					PACKAGE_NAME,
					"--stdio"
				]
			}
		},
		{
			name: "VS Code",
			configPath: node_path.default.join(baseDir, "Code", "User", "mcp.json"),
			configKey: "servers",
			format: "json",
			serverConfig: {
				type: "stdio",
				...stdioConfig
			}
		},
		{
			name: "Amp",
			configPath: node_path.default.join(homeDir, ".config", "amp", "settings.json"),
			configKey: "amp.mcpServers",
			format: "json",
			serverConfig: stdioConfig
		},
		{
			name: "Droid",
			configPath: node_path.default.join(homeDir, ".factory", "mcp.json"),
			configKey: "mcpServers",
			format: "json",
			serverConfig: {
				type: "stdio",
				...stdioConfig
			}
		},
		{
			name: "Windsurf",
			configPath: node_path.default.join(homeDir, ".codeium", "windsurf", "mcp_config.json"),
			configKey: "mcpServers",
			format: "json",
			serverConfig: stdioConfig
		},
		{
			name: "Zed",
			configPath: getZedConfigPath(),
			configKey: "context_servers",
			format: "json",
			serverConfig: {
				source: "custom",
				...stdioConfig,
				env: {}
			}
		}
	];
};
const ensureDirectory = (filePath) => {
	const directory = node_path.default.dirname(filePath);
	if (!node_fs.default.existsSync(directory)) node_fs.default.mkdirSync(directory, { recursive: true });
};
const JSONC_FORMAT_OPTIONS = {
	tabSize: 2,
	insertSpaces: true
};
const upsertIntoJsonc = (filePath, content, configKey, serverName, serverConfig) => {
	const edits = jsonc_parser.modify(content, [configKey, serverName], serverConfig, { formattingOptions: JSONC_FORMAT_OPTIONS });
	node_fs.default.writeFileSync(filePath, jsonc_parser.applyEdits(content, edits));
};
const installJsonClient = (client) => {
	ensureDirectory(client.configPath);
	const content = node_fs.default.existsSync(client.configPath) ? node_fs.default.readFileSync(client.configPath, "utf8") : "{}";
	upsertIntoJsonc(client.configPath, content, client.configKey, SERVER_NAME, client.serverConfig);
};
const installTomlClient = (client) => {
	ensureDirectory(client.configPath);
	const existingConfig = node_fs.default.existsSync(client.configPath) ? smol_toml.parse(node_fs.default.readFileSync(client.configPath, "utf8")) : {};
	const serverSection = existingConfig[client.configKey] ?? {};
	serverSection[SERVER_NAME] = client.serverConfig;
	existingConfig[client.configKey] = serverSection;
	node_fs.default.writeFileSync(client.configPath, smol_toml.stringify(existingConfig));
};
const getMcpClientNames = () => getClients().map((client) => client.name);
const installMcpServers = (selectedClients) => {
	const allClients = getClients();
	const clients = selectedClients ? allClients.filter((client) => selectedClients.includes(client.name)) : allClients;
	const results = [];
	const installSpinner = spinner("Installing MCP server.").start();
	for (const client of clients) try {
		if (client.format === "toml") installTomlClient(client);
		else installJsonClient(client);
		results.push({
			client: client.name,
			configPath: client.configPath,
			success: true
		});
	} catch (error) {
		const message = error instanceof Error ? error.message : String(error);
		results.push({
			client: client.name,
			configPath: client.configPath,
			success: false,
			error: message
		});
	}
	const successCount = results.filter((result) => result.success).length;
	if (successCount < results.length) installSpinner.warn(`Installed to ${successCount}/${results.length} agents.`);
	else installSpinner.succeed(`Installed to ${successCount} agents.`);
	for (const result of results) if (result.success) logger.log(`  ${highlighter.success("✓")} ${result.client} ${highlighter.dim("→")} ${highlighter.dim(result.configPath)}`);
	else logger.log(`  ${highlighter.error("✗")} ${result.client} ${highlighter.dim("→")} ${result.error}`);
	return results;
};
const promptMcpInstall = async () => {
	const { selectedAgents } = await prompts$1({
		type: "multiselect",
		name: "selectedAgents",
		message: "Select agents to install MCP server for:",
		choices: getMcpClientNames().map((name) => ({
			title: name,
			value: name,
			selected: true
		}))
	});
	if (selectedAgents === void 0 || selectedAgents.length === 0) return false;
	logger.break();
	return installMcpServers(selectedAgents).some((result) => result.success);
};
//#endregion
//#region src/commands/add.ts
const VERSION$5 = "0.1.31";
const add = new commander.Command().name("add").alias("install").description("connect React Grab to your agent via MCP").argument("[agent]", "agent to connect (mcp)").option("-y, --yes", "skip confirmation prompts", false).option("-c, --cwd <cwd>", "working directory (defaults to current directory)", process.cwd()).action(async (agentArg, opts) => {
	console.log(`${picocolors.default.magenta("✿")} ${picocolors.default.bold("React Grab")} ${picocolors.default.gray(VERSION$5)}`);
	console.log();
	try {
		const cwd = opts.cwd;
		const isNonInteractive = detectNonInteractive(opts.yes);
		const preflightSpinner = spinner("Preflight checks.").start();
		if (!(await detectProject(cwd)).hasReactGrab) {
			preflightSpinner.fail("React Grab is not installed.");
			logger.break();
			logger.error(`Run ${highlighter.info("react-grab init")} first to install React Grab.`);
			logger.break();
			process.exit(1);
		}
		preflightSpinner.succeed();
		if (agentArg && agentArg !== "mcp") {
			logger.break();
			logger.warn(`Legacy agent packages are deprecated. Use ${highlighter.info("mcp")} instead.`);
			logger.log(`Run ${highlighter.info("grab add mcp")} to install the MCP server.`);
			logger.break();
			process.exit(1);
		}
		if (agentArg === "mcp" || isNonInteractive) {
			if (isNonInteractive) {
				if (!installMcpServers().some((result) => result.success)) {
					logger.break();
					logger.error("Failed to install MCP server.");
					logger.break();
					process.exit(1);
				}
			} else if (!await promptMcpInstall()) {
				logger.break();
				process.exit(0);
			}
			logger.break();
			logger.log(`${highlighter.success("Success!")} MCP server has been configured.`);
			logger.log("Restart your agents to activate.");
			logger.break();
		} else {
			if (!await promptMcpInstall()) {
				logger.break();
				process.exit(0);
			}
			logger.break();
			logger.log(`${highlighter.success("Success!")} MCP server has been configured.`);
			logger.log("Restart your agents to activate.");
			logger.break();
		}
	} catch (error) {
		handleError(error);
	}
});
//#endregion
//#region src/utils/diff.ts
const RED = "\x1B[31m";
const GREEN = "\x1B[32m";
const GRAY = "\x1B[90m";
const RESET = "\x1B[0m";
const BOLD = "\x1B[1m";
const generateDiff = (originalContent, newContent) => {
	const originalLines = originalContent.split("\n");
	const newLines = newContent.split("\n");
	const diff = [];
	let originalIndex = 0;
	let newIndex = 0;
	while (originalIndex < originalLines.length || newIndex < newLines.length) {
		const originalLine = originalLines[originalIndex];
		const newLine = newLines[newIndex];
		if (originalLine === newLine) {
			diff.push({
				type: "unchanged",
				content: originalLine,
				lineNumber: newIndex + 1
			});
			originalIndex++;
			newIndex++;
		} else if (originalLine === void 0) {
			diff.push({
				type: "added",
				content: newLine,
				lineNumber: newIndex + 1
			});
			newIndex++;
		} else if (newLine === void 0) {
			diff.push({
				type: "removed",
				content: originalLine
			});
			originalIndex++;
		} else {
			const originalInNew = newLines.indexOf(originalLine, newIndex);
			const newInOriginal = originalLines.indexOf(newLine, originalIndex);
			if (originalInNew !== -1 && (newInOriginal === -1 || originalInNew - newIndex < newInOriginal - originalIndex)) while (newIndex < originalInNew) {
				diff.push({
					type: "added",
					content: newLines[newIndex],
					lineNumber: newIndex + 1
				});
				newIndex++;
			}
			else if (newInOriginal !== -1) while (originalIndex < newInOriginal) {
				diff.push({
					type: "removed",
					content: originalLines[originalIndex]
				});
				originalIndex++;
			}
			else {
				diff.push({
					type: "removed",
					content: originalLine
				});
				diff.push({
					type: "added",
					content: newLine,
					lineNumber: newIndex + 1
				});
				originalIndex++;
				newIndex++;
			}
		}
	}
	return diff;
};
const formatDiff = (diff, contextLines = 3) => {
	const lines = [];
	let lastPrintedIndex = -1;
	let hasChanges = false;
	const changedIndices = diff.map((line, index) => line.type !== "unchanged" ? index : -1).filter((index) => index !== -1);
	if (changedIndices.length === 0) return `${GRAY}No changes${RESET}`;
	for (const changedIndex of changedIndices) {
		const startContext = Math.max(0, changedIndex - contextLines);
		const endContext = Math.min(diff.length - 1, changedIndex + contextLines);
		if (startContext > lastPrintedIndex + 1 && lastPrintedIndex !== -1) lines.push(`${GRAY}  ...${RESET}`);
		for (let lineIndex = Math.max(startContext, lastPrintedIndex + 1); lineIndex <= endContext; lineIndex++) {
			const diffLine = diff[lineIndex];
			if (diffLine.type === "added") {
				lines.push(`${GREEN}+ ${diffLine.content}${RESET}`);
				hasChanges = true;
			} else if (diffLine.type === "removed") {
				lines.push(`${RED}- ${diffLine.content}${RESET}`);
				hasChanges = true;
			} else lines.push(`${GRAY}  ${diffLine.content}${RESET}`);
			lastPrintedIndex = lineIndex;
		}
	}
	return hasChanges ? lines.join("\n") : `${GRAY}No changes${RESET}`;
};
const printDiff = (filePath, originalContent, newContent) => {
	console.log(`\n${BOLD}File: ${filePath}${RESET}`);
	console.log("─".repeat(60));
	const diff = generateDiff(originalContent, newContent);
	console.log(formatDiff(diff));
	console.log("─".repeat(60));
};
//#endregion
//#region src/utils/templates.ts
const NEXT_APP_ROUTER_SCRIPT = `{process.env.NODE_ENV === "development" && (
          <Script
            src="//unpkg.com/react-grab/dist/index.global.js"
            crossOrigin="anonymous"
            strategy="beforeInteractive"
          />
        )}`;
const VITE_IMPORT = `if (import.meta.env.DEV) {
  import("react-grab");
}`;
const WEBPACK_IMPORT = `if (process.env.NODE_ENV === "development") {
  import("react-grab");
}`;
const TANSTACK_EFFECT = `useEffect(() => {
    if (import.meta.env.DEV) {
      void import("react-grab");
    }
  }, []);`;
const SCRIPT_IMPORT = "import Script from \"next/script\";";
//#endregion
//#region src/utils/transform.ts
const hasReactGrabCode = (content) => {
	return [
		/["'`][^"'`]*react-grab/,
		/react-grab[^"'`]*["'`]/,
		/<[^>]*react-grab/i,
		/import[^;]*react-grab/i,
		/require[^)]*react-grab/i,
		/from\s+[^;]*react-grab/i,
		/src[^>]*react-grab/i,
		/href[^>]*react-grab/i
	].some((pattern) => pattern.test(content));
};
const findLayoutFile = (projectRoot) => {
	const possiblePaths = [
		(0, node_path.join)(projectRoot, "app", "layout.tsx"),
		(0, node_path.join)(projectRoot, "app", "layout.jsx"),
		(0, node_path.join)(projectRoot, "src", "app", "layout.tsx"),
		(0, node_path.join)(projectRoot, "src", "app", "layout.jsx")
	];
	for (const filePath of possiblePaths) if ((0, node_fs.existsSync)(filePath)) return filePath;
	return null;
};
const findInstrumentationFile = (projectRoot) => {
	const possiblePaths = [
		(0, node_path.join)(projectRoot, "instrumentation-client.ts"),
		(0, node_path.join)(projectRoot, "instrumentation-client.js"),
		(0, node_path.join)(projectRoot, "src", "instrumentation-client.ts"),
		(0, node_path.join)(projectRoot, "src", "instrumentation-client.js")
	];
	for (const filePath of possiblePaths) if ((0, node_fs.existsSync)(filePath)) return filePath;
	return null;
};
const hasReactGrabInInstrumentation = (projectRoot) => {
	const instrumentationPath = findInstrumentationFile(projectRoot);
	if (!instrumentationPath) return false;
	return hasReactGrabCode((0, node_fs.readFileSync)(instrumentationPath, "utf-8"));
};
const findDocumentFile = (projectRoot) => {
	const possiblePaths = [
		(0, node_path.join)(projectRoot, "pages", "_document.tsx"),
		(0, node_path.join)(projectRoot, "pages", "_document.jsx"),
		(0, node_path.join)(projectRoot, "src", "pages", "_document.tsx"),
		(0, node_path.join)(projectRoot, "src", "pages", "_document.jsx")
	];
	for (const filePath of possiblePaths) if ((0, node_fs.existsSync)(filePath)) return filePath;
	return null;
};
const findIndexHtml = (projectRoot) => {
	const possiblePaths = [(0, node_path.join)(projectRoot, "index.html"), (0, node_path.join)(projectRoot, "public", "index.html")];
	for (const filePath of possiblePaths) if ((0, node_fs.existsSync)(filePath)) return filePath;
	return null;
};
const findEntryFile = (projectRoot) => {
	const possiblePaths = [
		(0, node_path.join)(projectRoot, "src", "index.tsx"),
		(0, node_path.join)(projectRoot, "src", "index.jsx"),
		(0, node_path.join)(projectRoot, "src", "index.ts"),
		(0, node_path.join)(projectRoot, "src", "index.js"),
		(0, node_path.join)(projectRoot, "src", "main.tsx"),
		(0, node_path.join)(projectRoot, "src", "main.jsx"),
		(0, node_path.join)(projectRoot, "src", "main.ts"),
		(0, node_path.join)(projectRoot, "src", "main.js")
	];
	for (const filePath of possiblePaths) if ((0, node_fs.existsSync)(filePath)) return filePath;
	return null;
};
const findTanStackRootFile = (projectRoot) => {
	const possiblePaths = [
		(0, node_path.join)(projectRoot, "src", "routes", "__root.tsx"),
		(0, node_path.join)(projectRoot, "src", "routes", "__root.jsx"),
		(0, node_path.join)(projectRoot, "app", "routes", "__root.tsx"),
		(0, node_path.join)(projectRoot, "app", "routes", "__root.jsx")
	];
	for (const filePath of possiblePaths) if ((0, node_fs.existsSync)(filePath)) return filePath;
	return null;
};
const alreadyConfiguredResult = (filePath) => ({
	success: true,
	filePath,
	message: "React Grab is already configured",
	noChanges: true
});
const transformNextAppRouter = (projectRoot, reactGrabAlreadyConfigured, force = false) => {
	const layoutPath = findLayoutFile(projectRoot);
	if (!layoutPath) return {
		success: false,
		filePath: "",
		message: "Could not find app/layout.tsx or app/layout.jsx"
	};
	const originalContent = (0, node_fs.readFileSync)(layoutPath, "utf-8");
	let newContent = originalContent;
	const hasReactGrabInFile = hasReactGrabCode(originalContent);
	const hasReactGrabInInstrumentationFile = hasReactGrabInInstrumentation(projectRoot);
	if (!force && hasReactGrabInFile && reactGrabAlreadyConfigured) return alreadyConfiguredResult(layoutPath);
	if (!force && (hasReactGrabInFile || hasReactGrabInInstrumentationFile)) return {
		success: true,
		filePath: layoutPath,
		message: "React Grab is already installed" + (hasReactGrabInInstrumentationFile ? " in instrumentation-client" : " in this file"),
		noChanges: true
	};
	if (!newContent.includes("import Script from \"next/script\"")) {
		const importMatch = newContent.match(/^import .+ from ['"].+['"];?\s*$/m);
		if (importMatch) newContent = newContent.replace(importMatch[0], `${importMatch[0]}\n${SCRIPT_IMPORT}`);
		else newContent = `${SCRIPT_IMPORT}\n\n${newContent}`;
	}
	const headMatch = newContent.match(/<head[^>]*>/);
	if (headMatch) newContent = newContent.replace(headMatch[0], `${headMatch[0]}\n        ${NEXT_APP_ROUTER_SCRIPT}`);
	else {
		const htmlMatch = newContent.match(/<html[^>]*>/);
		if (htmlMatch) newContent = newContent.replace(htmlMatch[0], `${htmlMatch[0]}\n      <head>\n        ${NEXT_APP_ROUTER_SCRIPT}\n      </head>`);
	}
	return {
		success: true,
		filePath: layoutPath,
		message: "Add React Grab",
		originalContent,
		newContent
	};
};
const transformNextPagesRouter = (projectRoot, reactGrabAlreadyConfigured, force = false) => {
	const documentPath = findDocumentFile(projectRoot);
	if (!documentPath) return {
		success: false,
		filePath: "",
		message: "Could not find pages/_document.tsx or pages/_document.jsx.\n\nTo set up React Grab with Pages Router, create pages/_document.tsx with:\n\n  import { Html, Head, Main, NextScript } from \"next/document\";\n  import Script from \"next/script\";\n\n  export default function Document() {\n    return (\n      <Html>\n        <Head>\n          {process.env.NODE_ENV === \"development\" && (\n            <Script src=\"//unpkg.com/react-grab/dist/index.global.js\" strategy=\"beforeInteractive\" />\n          )}\n        </Head>\n        <body>\n          <Main />\n          <NextScript />\n        </body>\n      </Html>\n    );\n  }"
	};
	const originalContent = (0, node_fs.readFileSync)(documentPath, "utf-8");
	let newContent = originalContent;
	const hasReactGrabInFile = hasReactGrabCode(originalContent);
	const hasReactGrabInInstrumentationFile = hasReactGrabInInstrumentation(projectRoot);
	if (!force && hasReactGrabInFile && reactGrabAlreadyConfigured) return alreadyConfiguredResult(documentPath);
	if (!force && (hasReactGrabInFile || hasReactGrabInInstrumentationFile)) return {
		success: true,
		filePath: documentPath,
		message: "React Grab is already installed" + (hasReactGrabInInstrumentationFile ? " in instrumentation-client" : " in this file"),
		noChanges: true
	};
	if (!newContent.includes("import Script from \"next/script\"")) {
		const importMatch = newContent.match(/^import .+ from ['"].+['"];?\s*$/m);
		if (importMatch) newContent = newContent.replace(importMatch[0], `${importMatch[0]}\n${SCRIPT_IMPORT}`);
	}
	const headMatch = newContent.match(/<Head[^>]*>/);
	if (headMatch) newContent = newContent.replace(headMatch[0], `${headMatch[0]}\n        ${NEXT_APP_ROUTER_SCRIPT}`);
	return {
		success: true,
		filePath: documentPath,
		message: "Add React Grab",
		originalContent,
		newContent
	};
};
const checkExistingInstallation = (filePath, reactGrabAlreadyConfigured) => {
	if (!hasReactGrabCode((0, node_fs.readFileSync)(filePath, "utf-8"))) return null;
	return {
		success: true,
		filePath,
		message: reactGrabAlreadyConfigured ? "React Grab is already configured" : "React Grab is already installed in this file",
		noChanges: true
	};
};
const transformVite = (projectRoot, reactGrabAlreadyConfigured, force = false) => {
	const entryPath = findEntryFile(projectRoot);
	if (!force) {
		const indexPath = findIndexHtml(projectRoot);
		if (indexPath) {
			const existingResult = checkExistingInstallation(indexPath, reactGrabAlreadyConfigured);
			if (existingResult) return existingResult;
		}
	}
	if (!entryPath) return {
		success: false,
		filePath: "",
		message: "Could not find entry file (src/index.tsx, src/main.tsx, etc.)"
	};
	if (!force) {
		const existingResult = checkExistingInstallation(entryPath, reactGrabAlreadyConfigured);
		if (existingResult) return existingResult;
	}
	const originalContent = (0, node_fs.readFileSync)(entryPath, "utf-8");
	return {
		success: true,
		filePath: entryPath,
		message: "Add React Grab",
		originalContent,
		newContent: `${VITE_IMPORT}\n\n${originalContent}`
	};
};
const transformWebpack = (projectRoot, reactGrabAlreadyConfigured, force = false) => {
	const entryPath = findEntryFile(projectRoot);
	if (!entryPath) return {
		success: false,
		filePath: "",
		message: "Could not find entry file (src/index.tsx, src/main.tsx, etc.)"
	};
	if (!force) {
		const existingResult = checkExistingInstallation(entryPath, reactGrabAlreadyConfigured);
		if (existingResult) return existingResult;
	}
	const originalContent = (0, node_fs.readFileSync)(entryPath, "utf-8");
	return {
		success: true,
		filePath: entryPath,
		message: "Add React Grab",
		originalContent,
		newContent: `${WEBPACK_IMPORT}\n\n${originalContent}`
	};
};
const transformTanStack = (projectRoot, reactGrabAlreadyConfigured, force = false) => {
	const rootPath = findTanStackRootFile(projectRoot);
	if (!rootPath) return {
		success: false,
		filePath: "",
		message: "Could not find src/routes/__root.tsx or app/routes/__root.tsx.\n\nTo set up React Grab with TanStack Start, add this to your root route component:\n\n  import { useEffect } from \"react\";\n\n  useEffect(() => {\n    if (import.meta.env.DEV) {\n      void import(\"react-grab\");\n    }\n  }, []);"
	};
	const originalContent = (0, node_fs.readFileSync)(rootPath, "utf-8");
	let newContent = originalContent;
	const hasReactGrabInFile = hasReactGrabCode(originalContent);
	if (!force && hasReactGrabInFile && reactGrabAlreadyConfigured) return alreadyConfiguredResult(rootPath);
	if (!force && hasReactGrabInFile) return {
		success: true,
		filePath: rootPath,
		message: "React Grab is already installed in this file",
		noChanges: true
	};
	if (!/import\s+\{[^}]*useEffect[^}]*\}\s+from\s+["']react["']/.test(newContent)) {
		const reactImportMatch = newContent.match(/import\s+\{([^}]*)\}\s+from\s+["']react["'];?/);
		if (reactImportMatch) {
			const existingImports = reactImportMatch[1];
			newContent = newContent.replace(reactImportMatch[0], `import { ${existingImports.trim()}, useEffect } from "react";`);
		} else {
			const firstImportMatch = newContent.match(/^import .+ from ['"].+['"];?\s*$/m);
			if (firstImportMatch) newContent = newContent.replace(firstImportMatch[0], `import { useEffect } from "react";\n${firstImportMatch[0]}`);
			else newContent = `import { useEffect } from "react";\n\n${newContent}`;
		}
	}
	const componentMatch = newContent.match(/function\s+(\w+)\s*\([^)]*\)\s*\{/);
	if (componentMatch) {
		const insertPosition = componentMatch.index + componentMatch[0].length;
		newContent = newContent.slice(0, insertPosition) + `\n  ${TANSTACK_EFFECT}\n` + newContent.slice(insertPosition);
	} else return {
		success: false,
		filePath: rootPath,
		message: "Could not find a component function in the root file"
	};
	return {
		success: true,
		filePath: rootPath,
		message: "Add React Grab",
		originalContent,
		newContent
	};
};
const previewTransform = (projectRoot, framework, nextRouterType, reactGrabAlreadyConfigured = false, force = false) => {
	switch (framework) {
		case "next":
			if (nextRouterType === "app") return transformNextAppRouter(projectRoot, reactGrabAlreadyConfigured, force);
			return transformNextPagesRouter(projectRoot, reactGrabAlreadyConfigured, force);
		case "vite": return transformVite(projectRoot, reactGrabAlreadyConfigured, force);
		case "tanstack": return transformTanStack(projectRoot, reactGrabAlreadyConfigured, force);
		case "webpack": return transformWebpack(projectRoot, reactGrabAlreadyConfigured, force);
		default: return {
			success: false,
			filePath: "",
			message: `Unknown framework: ${framework}. Please add React Grab manually.`
		};
	}
};
const canWriteToFile = (filePath) => {
	try {
		(0, node_fs.accessSync)(filePath, node_fs.constants.W_OK);
		return true;
	} catch {
		return false;
	}
};
const applyTransform = (result) => {
	if (result.success && result.newContent && result.filePath) {
		if (!canWriteToFile(result.filePath)) return {
			success: false,
			error: `Cannot write to ${result.filePath}. Check file permissions.`
		};
		try {
			(0, node_fs.writeFileSync)(result.filePath, result.newContent);
			return { success: true };
		} catch (error) {
			return {
				success: false,
				error: `Failed to write to ${result.filePath}: ${error instanceof Error ? error.message : "Unknown error"}`
			};
		}
	}
	return { success: true };
};
const formatOptionsForNextjs = (options) => {
	const parts = [];
	if (options.activationKey) parts.push(`activationKey: ${JSON.stringify(options.activationKey)}`);
	if (options.activationMode) parts.push(`activationMode: "${options.activationMode}"`);
	if (options.keyHoldDuration !== void 0) parts.push(`keyHoldDuration: ${options.keyHoldDuration}`);
	if (options.allowActivationInsideInput !== void 0) parts.push(`allowActivationInsideInput: ${options.allowActivationInsideInput}`);
	if (options.maxContextLines !== void 0) parts.push(`maxContextLines: ${options.maxContextLines}`);
	return `{ ${parts.join(", ")} }`;
};
const formatOptionsAsJson = (options) => {
	const cleanOptions = {};
	if (options.activationKey) cleanOptions.activationKey = options.activationKey;
	if (options.activationMode) cleanOptions.activationMode = options.activationMode;
	if (options.keyHoldDuration !== void 0) cleanOptions.keyHoldDuration = options.keyHoldDuration;
	if (options.allowActivationInsideInput !== void 0) cleanOptions.allowActivationInsideInput = options.allowActivationInsideInput;
	if (options.maxContextLines !== void 0) cleanOptions.maxContextLines = options.maxContextLines;
	return JSON.stringify(cleanOptions);
};
const findReactGrabFile = (projectRoot, framework, nextRouterType) => {
	switch (framework) {
		case "next":
			if (nextRouterType === "app") return findLayoutFile(projectRoot);
			return findDocumentFile(projectRoot);
		case "vite": {
			const entryFile = findEntryFile(projectRoot);
			if (entryFile && hasReactGrabCode((0, node_fs.readFileSync)(entryFile, "utf-8"))) return entryFile;
			const indexHtml = findIndexHtml(projectRoot);
			if (indexHtml && hasReactGrabCode((0, node_fs.readFileSync)(indexHtml, "utf-8"))) return indexHtml;
			return entryFile;
		}
		case "tanstack": return findTanStackRootFile(projectRoot);
		case "webpack": return findEntryFile(projectRoot);
		default: return null;
	}
};
const addOptionsToNextScript = (originalContent, options, filePath) => {
	const reactGrabScriptMatch = originalContent.match(/(<Script[\s\S]*?react-grab[\s\S]*?)\s*(\/?>)/i);
	if (!reactGrabScriptMatch) return {
		success: false,
		filePath,
		message: "Could not find React Grab Script tag"
	};
	const scriptTag = reactGrabScriptMatch[0];
	const scriptOpening = reactGrabScriptMatch[1];
	const scriptClosing = reactGrabScriptMatch[2];
	const existingDataOptionsMatch = scriptTag.match(/data-options=\{JSON\.stringify\([^)]+\)\}/);
	const dataOptionsAttr = `data-options={JSON.stringify(\n              ${formatOptionsForNextjs(options)}\n            )}`;
	let newScriptTag;
	if (existingDataOptionsMatch) newScriptTag = scriptTag.replace(existingDataOptionsMatch[0], dataOptionsAttr);
	else newScriptTag = `${scriptOpening}\n            ${dataOptionsAttr}\n          ${scriptClosing}`;
	return {
		success: true,
		filePath,
		message: "Update React Grab options",
		originalContent,
		newContent: originalContent.replace(scriptTag, newScriptTag)
	};
};
const addOptionsToDynamicImport = (originalContent, options, filePath) => {
	const reactGrabImportWithInitMatch = originalContent.match(/import\s*\(\s*["']react-grab["']\s*\)(?:\.then\s*\(\s*\(m\)\s*=>\s*m\.init\s*\([^)]*\)\s*\))?/);
	if (!reactGrabImportWithInitMatch) return {
		success: false,
		filePath,
		message: "Could not find React Grab import"
	};
	const newImport = `import("react-grab").then((m) => m.init(${formatOptionsAsJson(options)}))`;
	return {
		success: true,
		filePath,
		message: "Update React Grab options",
		originalContent,
		newContent: originalContent.replace(reactGrabImportWithInitMatch[0], newImport)
	};
};
const addOptionsToTanStackImport = (originalContent, options, filePath) => {
	const reactGrabImportWithInitMatch = originalContent.match(/(?:void\s+import\s*\(\s*["']react-grab["']\s*\)|import\s*\(\s*["']react-grab\/core["']\s*\)\.then\s*\(\s*\(\s*\{\s*init\s*\}\s*\)\s*=>\s*init\s*\([^)]*\)\s*\))/);
	if (!reactGrabImportWithInitMatch) return {
		success: false,
		filePath,
		message: "Could not find React Grab import"
	};
	const newImport = `import("react-grab/core").then(({ init }) => init(${formatOptionsAsJson(options)}))`;
	return {
		success: true,
		filePath,
		message: "Update React Grab options",
		originalContent,
		newContent: originalContent.replace(reactGrabImportWithInitMatch[0], newImport)
	};
};
const previewOptionsTransform = (projectRoot, framework, nextRouterType, options) => {
	const filePath = findReactGrabFile(projectRoot, framework, nextRouterType);
	if (!filePath) return {
		success: false,
		filePath: "",
		message: "Could not find file containing React Grab configuration"
	};
	const originalContent = (0, node_fs.readFileSync)(filePath, "utf-8");
	if (!hasReactGrabCode(originalContent)) return {
		success: false,
		filePath,
		message: "Could not find React Grab code in the file"
	};
	switch (framework) {
		case "next": return addOptionsToNextScript(originalContent, options, filePath);
		case "vite": return addOptionsToDynamicImport(originalContent, options, filePath);
		case "tanstack": return addOptionsToTanStackImport(originalContent, options, filePath);
		case "webpack": return addOptionsToDynamicImport(originalContent, options, filePath);
		default: return {
			success: false,
			filePath,
			message: `Unknown framework: ${framework}`
		};
	}
};
const previewCdnTransform = (projectRoot, framework, nextRouterType, targetCdnDomain) => {
	const filePath = findReactGrabFile(projectRoot, framework, nextRouterType);
	if (!filePath) return {
		success: false,
		filePath: "",
		message: "Could not find React Grab file"
	};
	const originalContent = (0, node_fs.readFileSync)(filePath, "utf-8");
	const newContent = originalContent.replace(/(https?:)?\/\/[^/\s"']+(?=\/(?:@?react-grab))/g, `//${targetCdnDomain}`).replace(/(https?:)?\/\/[^/\s"']*react-grab[^/\s"']*\.com(?=\/script\.js)/g, `//${targetCdnDomain}`);
	if (newContent === originalContent) return {
		success: true,
		filePath,
		message: "CDN already set",
		noChanges: true
	};
	return {
		success: true,
		filePath,
		message: "Update CDN",
		originalContent,
		newContent
	};
};
//#endregion
//#region src/utils/constants.ts
const MAX_KEY_HOLD_DURATION_MS = 2e3;
//#endregion
//#region src/utils/format-activation-key.ts
const formatActivationKeyDisplay = (activationKey) => {
	const defaultLabel = process.platform === "darwin" ? "Option" : "Alt";
	if (!activationKey) return `Default (${defaultLabel})`;
	return activationKey.split("+").map((part) => {
		const lower = part.toLowerCase();
		if (lower === "meta") return process.platform === "darwin" ? "⌘" : "Win";
		if (lower === "alt") return process.platform === "darwin" ? "⌥" : "Alt";
		if (lower === "ctrl") return "Ctrl";
		if (lower === "shift") return "Shift";
		if (lower === "space" || lower === " ") return "Space";
		return part.toUpperCase();
	}).join(" + ");
};
//#endregion
//#region src/commands/configure.ts
const VERSION$4 = "0.1.31";
const isMac = process.platform === "darwin";
const META_LABEL = isMac ? "Cmd" : "Win";
const ALT_LABEL = isMac ? "Option" : "Alt";
const MODIFIER_ALIASES = {
	cmd: "meta",
	command: "meta",
	win: "meta",
	windows: "meta",
	meta: "meta",
	ctrl: "ctrl",
	control: "ctrl",
	shift: "shift",
	alt: "alt",
	option: "alt",
	opt: "alt"
};
const MODIFIERS = [
	"meta",
	"ctrl",
	"shift",
	"alt"
];
const BASE_KEYS = [
	{
		key: " ",
		aliases: ["space", "spacebar"]
	},
	{
		key: "Enter",
		aliases: ["enter", "return"]
	},
	{
		key: "Escape",
		aliases: ["escape", "esc"]
	},
	{
		key: "Tab",
		aliases: ["tab"]
	},
	{
		key: "Backspace",
		aliases: ["backspace", "back"]
	},
	{
		key: "Delete",
		aliases: ["delete", "del"]
	},
	{
		key: "Insert",
		aliases: ["insert", "ins"]
	},
	{
		key: "Home",
		aliases: ["home"]
	},
	{
		key: "End",
		aliases: ["end"]
	},
	{
		key: "PageUp",
		aliases: ["pageup", "pgup"]
	},
	{
		key: "PageDown",
		aliases: [
			"pagedown",
			"pgdn",
			"pgdown"
		]
	},
	{
		key: "ArrowUp",
		aliases: ["arrowup", "up"]
	},
	{
		key: "ArrowDown",
		aliases: ["arrowdown", "down"]
	},
	{
		key: "ArrowLeft",
		aliases: ["arrowleft", "left"]
	},
	{
		key: "ArrowRight",
		aliases: ["arrowright", "right"]
	},
	...Array.from({ length: 12 }, (_, i) => ({
		key: `F${i + 1}`,
		aliases: [`f${i + 1}`]
	})),
	...Array.from({ length: 26 }, (_, i) => {
		const letter = String.fromCharCode(97 + i);
		return {
			key: letter,
			aliases: [letter]
		};
	}),
	...Array.from({ length: 10 }, (_, i) => ({
		key: String(i),
		aliases: [String(i)]
	})),
	{
		key: "`",
		aliases: [
			"backtick",
			"grave",
			"`"
		]
	},
	{
		key: "-",
		aliases: [
			"minus",
			"dash",
			"-"
		]
	},
	{
		key: "=",
		aliases: [
			"equals",
			"equal",
			"="
		]
	},
	{
		key: "[",
		aliases: [
			"leftbracket",
			"lbracket",
			"["
		]
	},
	{
		key: "]",
		aliases: [
			"rightbracket",
			"rbracket",
			"]"
		]
	},
	{
		key: "\\",
		aliases: ["backslash", "\\"]
	},
	{
		key: ";",
		aliases: ["semicolon", ";"]
	},
	{
		key: "'",
		aliases: [
			"quote",
			"apostrophe",
			"'"
		]
	},
	{
		key: ",",
		aliases: ["comma", ","]
	},
	{
		key: ".",
		aliases: [
			"period",
			"dot",
			"."
		]
	},
	{
		key: "/",
		aliases: [
			"slash",
			"forwardslash",
			"/"
		]
	}
];
const formatCombo = (combo) => {
	const parts = [];
	if (combo.metaKey) parts.push(META_LABEL);
	if (combo.ctrlKey) parts.push("Ctrl");
	if (combo.shiftKey) parts.push("Shift");
	if (combo.altKey) parts.push(ALT_LABEL);
	const keyDisplay = combo.key === " " ? "Space" : combo.key.length === 1 ? combo.key.toUpperCase() : combo.key;
	parts.push(keyDisplay);
	return parts.join("+");
};
const parseInput = (input) => {
	const parts = input.toLowerCase().replace(/\s+/g, "").split(/[+\-]/);
	const modifiers = /* @__PURE__ */ new Set();
	let partial = "";
	for (const part of parts) {
		if (!part) continue;
		const modifierKey = MODIFIER_ALIASES[part];
		if (modifierKey) modifiers.add(modifierKey);
		else partial = part;
	}
	return {
		modifiers,
		partial
	};
};
const POPULAR_KEYS = [
	"g",
	"k",
	"e",
	"d",
	"b",
	" ",
	"Escape",
	"Enter"
];
const generateSuggestions = (input) => {
	const { modifiers, partial } = parseInput(input);
	const suggestions = [];
	if (!partial && modifiers.size === 0 && !input) {
		for (const mod of MODIFIERS) {
			const label = mod === "meta" ? META_LABEL : mod === "alt" ? ALT_LABEL : mod.charAt(0).toUpperCase() + mod.slice(1);
			for (const popularKey of POPULAR_KEYS) {
				const keyDisplay = popularKey === " " ? "Space" : popularKey.length === 1 ? popularKey.toUpperCase() : popularKey;
				suggestions.push({
					title: `${label}+${keyDisplay}`,
					value: {
						key: popularKey,
						...mod === "meta" ? { metaKey: true } : {},
						...mod === "ctrl" ? { ctrlKey: true } : {},
						...mod === "shift" ? { shiftKey: true } : {},
						...mod === "alt" ? { altKey: true } : {}
					}
				});
			}
		}
		for (const baseKey of BASE_KEYS) suggestions.push({
			title: baseKey.key === " " ? "Space" : baseKey.key.length === 1 ? baseKey.key.toUpperCase() : baseKey.key,
			value: { key: baseKey.key }
		});
		return suggestions;
	}
	const buildCombo = (key, mods, extraMod) => ({
		key,
		...mods.has("meta") || extraMod === "meta" ? { metaKey: true } : {},
		...mods.has("ctrl") || extraMod === "ctrl" ? { ctrlKey: true } : {},
		...mods.has("shift") || extraMod === "shift" ? { shiftKey: true } : {},
		...mods.has("alt") || extraMod === "alt" ? { altKey: true } : {}
	});
	for (const baseKey of BASE_KEYS) if (partial ? baseKey.aliases.some((alias) => alias.startsWith(partial)) : true) {
		const combo = buildCombo(baseKey.key, modifiers);
		suggestions.push({
			title: formatCombo(combo),
			value: combo
		});
	}
	if (!partial) {
		const unusedMods = MODIFIERS.filter((m) => !modifiers.has(m));
		for (const mod of unusedMods) for (const popularKey of POPULAR_KEYS) {
			const combo = buildCombo(popularKey, modifiers, mod);
			suggestions.push({
				title: formatCombo(combo),
				value: combo
			});
		}
	}
	return suggestions.slice(0, 30);
};
const CONFIG_OPTIONS = [
	{
		id: "activationKey",
		title: "Activation Key",
		description: "The key used to activate React Grab (e.g., g, k, space)"
	},
	{
		id: "activationMode",
		title: "Activation Mode",
		description: "Toggle (press to activate/deactivate) or Hold (hold key)"
	},
	{
		id: "keyHoldDuration",
		title: "Key Hold Duration",
		description: "Milliseconds to hold the key before activation (hold mode)"
	},
	{
		id: "allowActivationInsideInput",
		title: "Allow Activation Inside Input",
		description: "Whether to allow activation when focused on input fields"
	},
	{
		id: "maxContextLines",
		title: "Max Context Lines",
		description: "Number of surrounding code lines to include in context"
	}
];
const comboToString = (combo) => {
	const parts = [];
	if (combo.metaKey) parts.push("Meta");
	if (combo.ctrlKey) parts.push("Ctrl");
	if (combo.shiftKey) parts.push("Shift");
	if (combo.altKey) parts.push("Alt");
	if (combo.key) {
		const keyDisplay = combo.key === " " ? "Space" : combo.key;
		parts.push(keyDisplay);
	}
	return parts.join("+");
};
const configure = new commander.Command().name("configure").alias("config").description("configure React Grab options").option("-y, --yes", "skip confirmation prompts", false).option("-k, --key <key>", "activation key (e.g., Meta+K, Ctrl+Shift+G, Space)").option("-m, --mode <mode>", "activation mode (toggle, hold)").option("--hold-duration <ms>", "key hold duration in milliseconds (for hold mode)").option("--allow-input <boolean>", "allow activation inside input fields (true/false)").option("--context-lines <lines>", "max context lines to include").option("--cdn <domain>", "CDN domain (e.g., unpkg.com, custom.react-grab.com)").option("-c, --cwd <cwd>", "working directory (defaults to current directory)", process.cwd()).action(async (opts) => {
	console.log(`${picocolors.default.magenta("✿")} ${picocolors.default.bold("React Grab")} ${picocolors.default.gray(VERSION$4)}`);
	console.log();
	try {
		const cwd = opts.cwd;
		const preflightSpinner = spinner("Preflight checks.").start();
		const projectInfo = await detectProject(cwd);
		if (!projectInfo.hasReactGrab) {
			preflightSpinner.fail("React Grab is not installed.");
			logger.break();
			logger.error(`Run ${highlighter.info("react-grab init")} first to install React Grab.`);
			logger.break();
			process.exit(1);
		}
		preflightSpinner.succeed();
		if (opts.cdn) {
			const result = previewCdnTransform(projectInfo.projectRoot, projectInfo.framework, projectInfo.nextRouterType, opts.cdn);
			if (!result.success) {
				logger.break();
				logger.error(result.message);
				logger.break();
				process.exit(1);
			}
			if (result.noChanges) {
				logger.break();
				logger.log("No changes needed.");
				logger.break();
				process.exit(0);
			}
			logger.break();
			printDiff(result.filePath, result.originalContent, result.newContent);
			if (!opts.yes) {
				logger.break();
				const { proceed } = await prompts$1({
					type: "confirm",
					name: "proceed",
					message: "Apply these changes?",
					initial: true
				});
				if (!proceed) {
					logger.break();
					logger.log("Changes cancelled.");
					logger.break();
					process.exit(0);
				}
			}
			const writeSpinner = spinner(`Applying changes to ${result.filePath}.`).start();
			const writeResult = applyTransform(result);
			if (!writeResult.success) {
				writeSpinner.fail();
				logger.break();
				logger.error(writeResult.error || "Failed to write file.");
				logger.break();
				process.exit(1);
			}
			writeSpinner.succeed();
			logger.break();
			logger.log(`${highlighter.success("Success!")} CDN updated.`);
			logger.break();
			return;
		}
		const hasFlags = opts.key || opts.mode || opts.holdDuration || opts.allowInput || opts.contextLines;
		logger.break();
		logger.log(`Configure ${highlighter.info("React Grab")} options:`);
		logger.break();
		const collectedOptions = {};
		if (hasFlags) {
			if (opts.key) {
				collectedOptions.activationKey = opts.key;
				logger.log(`  Activation key: ${highlighter.info(formatActivationKeyDisplay(collectedOptions.activationKey))}`);
			}
			if (opts.mode) {
				if (opts.mode !== "toggle" && opts.mode !== "hold") {
					logger.error(`Invalid mode: ${opts.mode}. Use "toggle" or "hold".`);
					logger.break();
					process.exit(1);
				}
				collectedOptions.activationMode = opts.mode;
				logger.log(`  Activation mode: ${highlighter.info(opts.mode)}`);
			}
			if (opts.holdDuration) {
				const duration = parseInt(opts.holdDuration, 10);
				if (isNaN(duration) || duration < 0 || duration > 2e3) {
					logger.error(`Invalid hold duration. Must be 0-${MAX_KEY_HOLD_DURATION_MS}ms.`);
					logger.break();
					process.exit(1);
				}
				collectedOptions.keyHoldDuration = duration;
				logger.log(`  Key hold duration: ${highlighter.info(`${duration}ms`)}`);
			}
			if (opts.allowInput !== void 0) {
				const allowInput = opts.allowInput === "true" || opts.allowInput === true;
				collectedOptions.allowActivationInsideInput = allowInput;
				logger.log(`  Allow activation inside input: ${highlighter.info(String(allowInput))}`);
			}
			if (opts.contextLines) {
				const lines = parseInt(opts.contextLines, 10);
				if (isNaN(lines) || lines < 0 || lines > 50) {
					logger.error(`Invalid context lines. Must be 0-50.`);
					logger.break();
					process.exit(1);
				}
				collectedOptions.maxContextLines = lines;
				logger.log(`  Max context lines: ${highlighter.info(String(lines))}`);
			}
		} else {
			const { selectedOption } = await prompts$1({
				type: "autocomplete",
				name: "selectedOption",
				message: "Search for an option to configure:",
				choices: CONFIG_OPTIONS.map((option) => ({
					title: option.title,
					value: option.id,
					description: option.description
				})),
				suggest: (input, choices) => Promise.resolve(choices.filter((choice) => choice.title.toLowerCase().includes(input.toLowerCase()) || (choice.description?.toLowerCase().includes(input.toLowerCase()) ?? false)))
			});
			if (selectedOption === void 0) {
				logger.break();
				process.exit(1);
			}
			if (selectedOption === "activationKey") {
				const { selectedCombo } = await prompts$1({
					type: "autocomplete",
					name: "selectedCombo",
					message: "Type key combination (e.g. ctrl+shift+g):",
					choices: generateSuggestions(""),
					suggest: (input) => Promise.resolve(generateSuggestions(input))
				});
				if (selectedCombo === void 0) {
					logger.break();
					process.exit(1);
				}
				collectedOptions.activationKey = comboToString(selectedCombo);
				logger.log(`  Activation key: ${highlighter.info(formatActivationKeyDisplay(collectedOptions.activationKey))}`);
			}
			if (selectedOption === "activationMode") {
				const { activationMode } = await prompts$1({
					type: "select",
					name: "activationMode",
					message: `Select ${highlighter.info("activation mode")}:`,
					choices: [{
						title: "Toggle (press to activate/deactivate)",
						value: "toggle"
					}, {
						title: "Hold (hold key to keep active)",
						value: "hold"
					}],
					initial: 0
				});
				if (activationMode === void 0) {
					logger.break();
					process.exit(1);
				}
				collectedOptions.activationMode = activationMode;
			}
			if (selectedOption === "keyHoldDuration") {
				const { keyHoldDuration } = await prompts$1({
					type: "number",
					name: "keyHoldDuration",
					message: `Enter ${highlighter.info("key hold duration")} in milliseconds:`,
					initial: 150,
					min: 0,
					max: 2e3
				});
				if (keyHoldDuration === void 0) {
					logger.break();
					process.exit(1);
				}
				collectedOptions.keyHoldDuration = keyHoldDuration;
			}
			if (selectedOption === "allowActivationInsideInput") {
				const { allowActivationInsideInput } = await prompts$1({
					type: "confirm",
					name: "allowActivationInsideInput",
					message: `Allow activation ${highlighter.info("inside input fields")}?`,
					initial: true
				});
				if (allowActivationInsideInput === void 0) {
					logger.break();
					process.exit(1);
				}
				collectedOptions.allowActivationInsideInput = allowActivationInsideInput;
			}
			if (selectedOption === "maxContextLines") {
				const { maxContextLines } = await prompts$1({
					type: "number",
					name: "maxContextLines",
					message: `Enter ${highlighter.info("max context lines")} to include:`,
					initial: 3,
					min: 0,
					max: 50
				});
				if (maxContextLines === void 0) {
					logger.break();
					process.exit(1);
				}
				collectedOptions.maxContextLines = maxContextLines;
			}
		}
		const result = previewOptionsTransform(projectInfo.projectRoot, projectInfo.framework, projectInfo.nextRouterType, collectedOptions);
		if (!result.success) {
			logger.break();
			logger.warn(result.message);
			logger.break();
			const configJson = JSON.stringify(collectedOptions);
			logger.log(`Add this to your ${highlighter.info("init()")} call or ${highlighter.info("data-options")} attribute:`);
			logger.break();
			console.log(`  ${picocolors.default.cyan(configJson)}`);
			logger.break();
			process.exit(1);
		}
		if (!result.noChanges && result.originalContent && result.newContent) {
			logger.break();
			printDiff(result.filePath, result.originalContent, result.newContent);
			if (!opts.yes) {
				logger.break();
				const { proceed } = await prompts$1({
					type: "confirm",
					name: "proceed",
					message: "Apply these changes?",
					initial: true
				});
				if (!proceed) {
					logger.break();
					logger.log("Changes cancelled.");
					logger.break();
					process.exit(0);
				}
			}
			const writeSpinner = spinner(`Applying changes to ${result.filePath}.`).start();
			const writeResult = applyTransform(result);
			if (!writeResult.success) {
				writeSpinner.fail();
				logger.break();
				logger.error(writeResult.error || "Failed to write file.");
				logger.break();
				process.exit(1);
			}
			writeSpinner.succeed();
		} else {
			logger.break();
			logger.log("No changes needed.");
		}
		logger.break();
		logger.log(`${highlighter.success("Success!")} React Grab options have been configured.`);
		logger.break();
	} catch (error) {
		handleError(error);
	}
});
//#endregion
//#region src/utils/install.ts
const INSTALL_COMMANDS = {
	npm: "npm install",
	yarn: "yarn add",
	pnpm: "pnpm add",
	bun: "bun add"
};
const installPackages = (packages, packageManager, projectRoot, isDev = true) => {
	if (packages.length === 0) return;
	const fullCommand = `${INSTALL_COMMANDS[packageManager]}${isDev ? " -D" : ""} ${packages.join(" ")}`;
	console.log(`Running: ${fullCommand}\n`);
	(0, node_child_process.execSync)(fullCommand, {
		cwd: projectRoot,
		stdio: "inherit",
		env: {
			...process.env,
			REACT_GRAB_INIT: "1"
		}
	});
};
const getPackagesToInstall = (includeReactGrab = true) => {
	return includeReactGrab ? ["react-grab"] : [];
};
//#endregion
//#region src/utils/cli-helpers.ts
const applyTransformWithFeedback = (result, message) => {
	const writeSpinner = spinner(message ?? `Applying changes to ${result.filePath}.`).start();
	const writeResult = applyTransform(result);
	if (!writeResult.success) {
		writeSpinner.fail();
		logger.break();
		logger.error(writeResult.error || "Failed to write file.");
		logger.break();
		process.exit(1);
	}
	writeSpinner.succeed();
};
const installPackagesWithFeedback = (packages, packageManager, projectRoot) => {
	if (packages.length === 0) return;
	const installSpinner = spinner(`Installing ${packages.join(", ")}.`).start();
	try {
		installPackages(packages, packageManager, projectRoot);
		installSpinner.succeed();
	} catch (error) {
		installSpinner.fail();
		handleError(error);
	}
};
//#endregion
//#region src/commands/init.ts
const VERSION$3 = "0.1.31";
const REPORT_URL = "https://react-grab.com/api/report-cli";
const DOCS_URL = "https://github.com/aidenybai/react-grab";
const reportToCli = (type, config, error) => {
	fetch(REPORT_URL, {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify({
			type,
			version: VERSION$3,
			config,
			error: error ? {
				message: error.message,
				stack: error.stack
			} : void 0,
			timestamp: (/* @__PURE__ */ new Date()).toISOString()
		})
	}).catch(() => {});
};
const FRAMEWORK_NAMES = {
	next: "Next.js",
	vite: "Vite",
	tanstack: "TanStack Start",
	webpack: "Webpack",
	unknown: "Unknown"
};
const PACKAGE_MANAGER_NAMES = {
	npm: "npm",
	yarn: "Yarn",
	pnpm: "pnpm",
	bun: "Bun"
};
const UNSUPPORTED_FRAMEWORK_NAMES = {
	remix: "Remix",
	astro: "Astro",
	sveltekit: "SvelteKit",
	gatsby: "Gatsby"
};
const sortProjectsByFramework = (projects) => [...projects].sort((projectA, projectB) => {
	if (projectA.framework === "unknown" && projectB.framework !== "unknown") return 1;
	if (projectA.framework !== "unknown" && projectB.framework === "unknown") return -1;
	return 0;
});
const printSubprojects = (searchRoot, sortedProjects) => {
	logger.break();
	logger.log("Found the following projects:");
	logger.break();
	for (const project of sortedProjects) {
		const frameworkLabel = project.framework !== "unknown" ? ` ${highlighter.dim(`(${FRAMEWORK_NAMES[project.framework]})`)}` : "";
		const relativePath = (0, node_path.relative)(searchRoot, project.path);
		logger.log(`  ${highlighter.info(project.name)}${frameworkLabel} ${highlighter.dim(relativePath)}`);
	}
	logger.break();
	logger.log(`Re-run with ${highlighter.info("-c <path>")} to specify a project:`);
	logger.break();
	logger.log(`  ${highlighter.dim("$")} npx grab@latest init -c ${(0, node_path.relative)(searchRoot, sortedProjects[0].path)}`);
	logger.break();
};
const init = new commander.Command().name("init").alias("setup").description("initialize React Grab in your project").option("-y, --yes", "skip confirmation prompts", false).option("-f, --force", "force overwrite existing config", false).option("-k, --key <key>", "activation key (e.g., Meta+K, Ctrl+Shift+G, Space)").option("--skip-install", "skip package installation", false).option("--pkg <pkg>", "custom package URL for CLI (e.g., grab)").option("-c, --cwd <cwd>", "working directory (defaults to current directory)", process.cwd()).action(async (opts) => {
	console.log(`${picocolors.default.magenta("✿")} ${picocolors.default.bold("React Grab")} ${picocolors.default.gray(VERSION$3)}`);
	console.log();
	try {
		const cwd = (0, node_path.resolve)(opts.cwd);
		const isNonInteractive = detectNonInteractive(opts.yes);
		if (!(0, node_fs.existsSync)(cwd)) {
			logger.break();
			logger.error(`Directory does not exist: ${highlighter.info(cwd)}`);
			logger.break();
			process.exit(1);
		}
		const preflightSpinner = spinner("Preflight checks.").start();
		const projectInfo = await detectProject(cwd);
		if (projectInfo.hasReactGrab && !opts.force) {
			preflightSpinner.succeed();
			if (isNonInteractive) {
				logger.break();
				logger.warn("React Grab is already installed.");
				logger.log(`Use ${highlighter.info("--force")} to reconfigure, or remove ${highlighter.info("--yes")} for interactive mode.`);
				logger.break();
				process.exit(0);
			}
			logger.break();
			logger.success("React Grab is already installed.");
			logger.break();
			const { wantCustomizeOptions } = await prompts$1({
				type: "confirm",
				name: "wantCustomizeOptions",
				message: `Would you like to customize ${highlighter.info("options")}?`,
				initial: false
			});
			if (wantCustomizeOptions === void 0) {
				logger.break();
				process.exit(1);
			}
			if (wantCustomizeOptions || opts.key) {
				logger.break();
				logger.log(`Configure ${highlighter.info("React Grab")} options:`);
				logger.break();
				const collectedOptions = {};
				if (opts.key) {
					collectedOptions.activationKey = opts.key;
					logger.log(`  Activation key: ${highlighter.info(formatActivationKeyDisplay(collectedOptions.activationKey))}`);
				} else {
					const { wantActivationKey } = await prompts$1({
						type: "confirm",
						name: "wantActivationKey",
						message: `Configure ${highlighter.info("activation key")}?`,
						initial: false
					});
					if (wantActivationKey === void 0) {
						logger.break();
						process.exit(1);
					}
					if (wantActivationKey) {
						const { key } = await prompts$1({
							type: "text",
							name: "key",
							message: "Enter the activation key (e.g., g, k, space):",
							initial: ""
						});
						if (key === void 0) {
							logger.break();
							process.exit(1);
						}
						collectedOptions.activationKey = key ? key.toLowerCase() : void 0;
						logger.log(`  Activation key: ${highlighter.info(formatActivationKeyDisplay(collectedOptions.activationKey))}`);
					}
				}
				const { activationMode } = await prompts$1({
					type: "select",
					name: "activationMode",
					message: `Select ${highlighter.info("activation mode")}:`,
					choices: [{
						title: "Toggle (press to activate/deactivate)",
						value: "toggle"
					}, {
						title: "Hold (hold key to keep active)",
						value: "hold"
					}],
					initial: 0
				});
				if (activationMode === void 0) {
					logger.break();
					process.exit(1);
				}
				collectedOptions.activationMode = activationMode;
				if (activationMode === "hold") {
					const { keyHoldDuration } = await prompts$1({
						type: "number",
						name: "keyHoldDuration",
						message: `Enter ${highlighter.info("key hold duration")} in milliseconds:`,
						initial: 150,
						min: 0,
						max: 2e3
					});
					if (keyHoldDuration === void 0) {
						logger.break();
						process.exit(1);
					}
					collectedOptions.keyHoldDuration = keyHoldDuration;
				}
				const { allowActivationInsideInput } = await prompts$1({
					type: "confirm",
					name: "allowActivationInsideInput",
					message: `Allow activation ${highlighter.info("inside input fields")}?`,
					initial: true
				});
				if (allowActivationInsideInput === void 0) {
					logger.break();
					process.exit(1);
				}
				collectedOptions.allowActivationInsideInput = allowActivationInsideInput;
				const { maxContextLines } = await prompts$1({
					type: "number",
					name: "maxContextLines",
					message: `Enter ${highlighter.info("max context lines")} to include:`,
					initial: 3,
					min: 0,
					max: 50
				});
				if (maxContextLines === void 0) {
					logger.break();
					process.exit(1);
				}
				collectedOptions.maxContextLines = maxContextLines;
				const optionsResult = previewOptionsTransform(projectInfo.projectRoot, projectInfo.framework, projectInfo.nextRouterType, collectedOptions);
				if (!optionsResult.success) {
					logger.break();
					logger.error(optionsResult.message);
					logger.break();
					process.exit(1);
				}
				if (!optionsResult.noChanges && optionsResult.originalContent && optionsResult.newContent) {
					logger.break();
					printDiff(optionsResult.filePath, optionsResult.originalContent, optionsResult.newContent);
					logger.break();
					const { proceed } = await prompts$1({
						type: "confirm",
						name: "proceed",
						message: "Apply these changes?",
						initial: true
					});
					if (!proceed) {
						logger.break();
						logger.log("Options configuration cancelled.");
					} else {
						applyTransformWithFeedback(optionsResult);
						logger.break();
						logger.success("React Grab options have been configured.");
					}
				} else {
					logger.break();
					logger.log("No option changes needed.");
				}
			}
			logger.break();
			const { wantAddMcp } = await prompts$1({
				type: "confirm",
				name: "wantAddMcp",
				message: `Would you like to ${highlighter.info("connect it to your agent via MCP")}?`,
				initial: false
			});
			if (wantAddMcp === void 0) {
				logger.break();
				process.exit(1);
			}
			if (wantAddMcp) {
				if (!await promptMcpInstall()) {
					logger.break();
					process.exit(0);
				}
				logger.break();
				logger.success("MCP server has been configured.");
				logger.log("Restart your agents to activate.");
			}
			logger.break();
			process.exit(0);
		}
		preflightSpinner.succeed();
		const frameworkSpinner = spinner("Verifying framework.").start();
		if (projectInfo.unsupportedFramework) {
			const frameworkName = UNSUPPORTED_FRAMEWORK_NAMES[projectInfo.unsupportedFramework];
			frameworkSpinner.fail(`Found ${highlighter.info(frameworkName)}.`);
			logger.break();
			logger.log(`${frameworkName} is not yet supported by automatic setup.`);
			logger.log(`Visit ${highlighter.info(DOCS_URL)} for manual setup.`);
			logger.break();
			process.exit(1);
		}
		if (projectInfo.framework === "unknown") {
			let searchRoot = cwd;
			let reactProjects = findReactProjects(searchRoot);
			if (reactProjects.length === 0 && cwd !== process.cwd()) {
				searchRoot = process.cwd();
				reactProjects = findReactProjects(searchRoot);
			}
			if (reactProjects.length > 0) {
				frameworkSpinner.info(`Verifying framework. Found ${reactProjects.length} project${reactProjects.length === 1 ? "" : "s"}.`);
				const sortedProjects = sortProjectsByFramework(reactProjects);
				if (isNonInteractive) {
					printSubprojects(searchRoot, sortedProjects);
					process.exit(1);
				}
				logger.break();
				const { selectedProject } = await prompts$1({
					type: "select",
					name: "selectedProject",
					message: "Select a project to install React Grab:",
					choices: [...sortedProjects.map((project) => {
						const frameworkLabel = project.framework !== "unknown" ? ` ${highlighter.dim(`(${FRAMEWORK_NAMES[project.framework]})`)}` : "";
						return {
							title: `${project.name}${frameworkLabel}`,
							value: project.path
						};
					}), {
						title: "Skip",
						value: "skip"
					}]
				});
				if (!selectedProject || selectedProject === "skip") {
					logger.break();
					process.exit(0);
				}
				process.chdir(selectedProject);
				const newProjectInfo = await detectProject(selectedProject);
				Object.assign(projectInfo, newProjectInfo);
				spinner("Verifying framework.").start().succeed(`Verifying framework. Found ${highlighter.info(FRAMEWORK_NAMES[newProjectInfo.framework])}.`);
			} else {
				frameworkSpinner.fail("Could not detect a supported framework.");
				logger.break();
				logger.log("React Grab supports Next.js, Vite, TanStack Start, and Webpack projects.");
				logger.log(`Visit ${highlighter.info(DOCS_URL)} for manual setup.`);
				logger.break();
				process.exit(1);
			}
		} else frameworkSpinner.succeed(`Verifying framework. Found ${highlighter.info(FRAMEWORK_NAMES[projectInfo.framework])}.`);
		if (projectInfo.framework === "next") spinner("Detecting router type.").start().succeed(`Detecting router type. Found ${highlighter.info(projectInfo.nextRouterType === "app" ? "App Router" : "Pages Router")}.`);
		spinner("Detecting package manager.").start().succeed(`Detecting package manager. Found ${highlighter.info(PACKAGE_MANAGER_NAMES[projectInfo.packageManager])}.`);
		const finalFramework = projectInfo.framework;
		const finalPackageManager = projectInfo.packageManager;
		const finalNextRouterType = projectInfo.nextRouterType;
		let didInstallMcp = false;
		if (!isNonInteractive) {
			logger.break();
			const { wantAddMcp } = await prompts$1({
				type: "confirm",
				name: "wantAddMcp",
				message: `Would you like to ${highlighter.info("connect it to your agent via MCP")}?`,
				initial: false
			});
			if (wantAddMcp === void 0) {
				logger.break();
				process.exit(1);
			}
			if (wantAddMcp) {
				didInstallMcp = Boolean(await promptMcpInstall());
				if (!didInstallMcp) {
					logger.break();
					process.exit(0);
				}
				logger.break();
				logger.success("MCP server has been configured.");
				logger.log("Continuing with React Grab installation...");
				logger.break();
			}
		}
		const result = previewTransform(projectInfo.projectRoot, finalFramework, finalNextRouterType, false, opts.force);
		if (!result.success) {
			logger.break();
			logger.error(result.message);
			logger.error(`Visit ${highlighter.info(DOCS_URL)} for manual setup.`);
			logger.break();
			process.exit(1);
		}
		const hasLayoutChanges = !result.noChanges && result.originalContent && result.newContent;
		if (hasLayoutChanges) {
			logger.break();
			printDiff(result.filePath, result.originalContent, result.newContent);
			logger.break();
			logger.warn("Auto-detection may not be 100% accurate.");
			logger.warn("Please verify the changes before committing.");
			if (!isNonInteractive) {
				logger.break();
				const { proceed } = await prompts$1({
					type: "confirm",
					name: "proceed",
					message: "Apply these changes?",
					initial: true
				});
				if (!proceed) {
					logger.break();
					logger.log("Changes cancelled.");
					logger.break();
					process.exit(0);
				}
			}
		}
		const shouldInstallReactGrab = !projectInfo.hasReactGrab;
		if (!opts.skipInstall && shouldInstallReactGrab) installPackagesWithFeedback(getPackagesToInstall(shouldInstallReactGrab), finalPackageManager, projectInfo.projectRoot);
		if (hasLayoutChanges) applyTransformWithFeedback(result);
		logger.break();
		logger.log(`${highlighter.success("Success!")} React Grab has been installed.`);
		logger.log("You may now start your development server.");
		logger.break();
		reportToCli("completed", {
			framework: finalFramework,
			packageManager: finalPackageManager,
			router: finalNextRouterType,
			agent: didInstallMcp ? "mcp" : void 0,
			isMonorepo: projectInfo.isMonorepo
		});
	} catch (error) {
		handleError(error);
		reportToCli("error", void 0, error);
	}
});
//#endregion
//#region src/commands/remove.ts
const VERSION$2 = "0.1.31";
const remove = new commander.Command().name("remove").description("disconnect React Grab from your agent").argument("[agent]", "agent to disconnect (mcp)").option("-y, --yes", "skip confirmation prompts", false).option("-c, --cwd <cwd>", "working directory (defaults to current directory)", process.cwd()).action(async (agentArg, opts) => {
	console.log(`${picocolors.default.magenta("✿")} ${picocolors.default.bold("React Grab")} ${picocolors.default.gray(VERSION$2)}`);
	console.log();
	try {
		const cwd = opts.cwd;
		const preflightSpinner = spinner("Preflight checks.").start();
		if (!(await detectProject(cwd)).hasReactGrab) {
			preflightSpinner.fail("React Grab is not installed.");
			logger.break();
			logger.error(`Run ${highlighter.info("react-grab init")} first to install React Grab.`);
			logger.break();
			process.exit(1);
		}
		preflightSpinner.succeed();
		if (agentArg && agentArg !== "mcp") {
			logger.break();
			logger.warn(`Legacy agent packages are deprecated. Uninstall ${highlighter.info(`@react-grab/${agentArg}`)} manually with your package manager.`);
			logger.break();
			process.exit(0);
		}
		logger.break();
		logger.warn("To remove the MCP server, delete the react-grab-mcp entry from your agent's MCP config file.");
		logger.break();
	} catch (error) {
		handleError(error);
	}
});
//#endregion
//#region src/commands/upgrade.ts
const VERSION$1 = "0.1.31";
const NPM_REGISTRY_URL = "https://registry.npmjs.org/react-grab/latest";
const fetchLatestVersion = async () => {
	try {
		return (await (await fetch(NPM_REGISTRY_URL)).json()).version ?? null;
	} catch {
		return null;
	}
};
const isDevDependency = (projectRoot) => {
	const packageJsonPath = (0, node_path.join)(projectRoot, "package.json");
	if (!(0, node_fs.existsSync)(packageJsonPath)) return true;
	try {
		const packageJson = JSON.parse((0, node_fs.readFileSync)(packageJsonPath, "utf-8"));
		if (packageJson.devDependencies?.["react-grab"]) return true;
		if (packageJson.dependencies?.["react-grab"]) return false;
	} catch {}
	return true;
};
const upgrade = new commander.Command().name("upgrade").alias("update").description("upgrade react-grab to the latest version").option("-c, --cwd <cwd>", "working directory (defaults to current directory)", process.cwd()).action(async (opts) => {
	console.log(`${picocolors.default.magenta("✿")} ${picocolors.default.bold("React Grab")} ${picocolors.default.gray(VERSION$1)}`);
	console.log();
	try {
		const cwd = (0, node_path.resolve)(opts.cwd);
		const detectSpinner = spinner("Detecting project.").start();
		const projectInfo = await detectProject(cwd);
		if (!projectInfo.hasReactGrab) {
			detectSpinner.fail("React Grab is not installed.");
			logger.break();
			logger.error(`Run ${highlighter.info("npx grab@latest init")} first to install React Grab.`);
			logger.break();
			process.exit(1);
		}
		detectSpinner.succeed();
		const versionSpinner = spinner("Checking for updates.").start();
		const latestVersion = await fetchLatestVersion();
		if (!latestVersion) {
			versionSpinner.fail("Could not check for updates.");
			logger.break();
			logger.error("Failed to reach the npm registry. Check your network connection.");
			logger.break();
			process.exit(1);
		}
		const installedVersion = projectInfo.reactGrabVersion;
		if (installedVersion && installedVersion === latestVersion) {
			versionSpinner.succeed(`Already on the latest version ${highlighter.info(`v${latestVersion}`)}.`);
			logger.break();
			process.exit(0);
		}
		const fromLabel = installedVersion ? `v${installedVersion}` : "unknown";
		versionSpinner.succeed(`Update available: ${highlighter.dim(fromLabel)} → ${highlighter.info(`v${latestVersion}`)}.`);
		const upgradeSpinner = spinner("Upgrading react-grab.").start();
		try {
			installPackages(["react-grab@latest"], projectInfo.packageManager, projectInfo.projectRoot, isDevDependency(projectInfo.projectRoot));
			upgradeSpinner.succeed();
		} catch {
			upgradeSpinner.fail();
			logger.break();
			logger.error("Failed to upgrade. Check your network connection and try again.");
			logger.break();
			process.exit(1);
		}
		logger.break();
		logger.log(`${highlighter.success("Success!")} React Grab has been upgraded to ${highlighter.info(`v${latestVersion}`)}.`);
		logger.break();
	} catch (error) {
		handleError(error);
	}
});
//#endregion
//#region src/cli.ts
const VERSION = "0.1.31";
const VERSION_API_URL = "https://www.react-grab.com/api/version";
process.on("SIGINT", () => process.exit(0));
process.on("SIGTERM", () => process.exit(0));
try {
	fetch(`${VERSION_API_URL}?source=cli&v=${VERSION}&t=${Date.now()}`).catch(() => {});
} catch {}
const program = new commander.Command().name("grab").description("add React Grab to your project").version(VERSION, "-v, --version", "display the version number");
program.addCommand(init);
program.addCommand(add);
program.addCommand(remove);
program.addCommand(configure);
program.addCommand(upgrade);
const main = async () => {
	await program.parseAsync();
};
main();
//#endregion
