// react-grab@0.1.31/dist/state-C7ZMJv5S.cjs
/**
 * @license MIT
 *
 * Copyright (c) 2025 Aiden Bai
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const require_freeze_updates = require('./freeze-updates-Dj9eqIro.cjs');

//#region src/utils/is-element-connected.ts
const isElementConnected = (element) => Boolean(element?.isConnected ?? element?.ownerDocument?.contains(element));

//#endregion
//#region src/utils/is-keyboard-event-triggered-by-input.ts
const EDITABLE_TAGS_AND_ROLES = [
	"input",
	"textarea",
	"select",
	"searchbox",
	"slider",
	"spinbutton",
	"menuitem",
	"menuitemcheckbox",
	"menuitemradio",
	"option",
	"radio",
	"textbox",
	"combobox"
];
const getTargetElement = (event) => {
	if (event.composed) {
		const firstElement = event.composedPath()[0];
		if (firstElement instanceof HTMLElement) return firstElement;
	} else if (event.target instanceof HTMLElement) return event.target;
};
const isKeyboardEventTriggeredByInput = (event) => {
	if (document.designMode === "on") return true;
	const targetElement = getTargetElement(event);
	if (!targetElement) return false;
	if (targetElement.isContentEditable) return true;
	const tagName = require_freeze_updates.getTagName(targetElement);
	return EDITABLE_TAGS_AND_ROLES.some((tagOrRole) => tagOrRole === tagName || tagOrRole === targetElement.role);
};
const hasTextSelectionInInput = (event) => {
	const target = event.target;
	if (target instanceof HTMLInputElement || target instanceof HTMLTextAreaElement) {
		const selectionStart = target.selectionStart ?? 0;
		return (target.selectionEnd ?? 0) - selectionStart > 0;
	}
	return false;
};
const hasTextSelectionOnPage = () => {
	const selection = window.getSelection();
	if (!selection) return false;
	return selection.toString().length > 0;
};

//#endregion
//#region src/utils/get-visual-viewport.ts
const getVisualViewport = () => {
	const visualViewport = window.visualViewport;
	if (visualViewport) return {
		width: visualViewport.width,
		height: visualViewport.height,
		offsetLeft: visualViewport.offsetLeft,
		offsetTop: visualViewport.offsetTop
	};
	return {
		width: window.innerWidth,
		height: window.innerHeight,
		offsetLeft: 0,
		offsetTop: 0
	};
};

//#endregion
//#region src/utils/toolbar-position.ts
const clampToRange = (value, min, max) => Math.max(min, Math.min(value, max));
const getPositionFromEdgeAndRatio = (edge, ratio, elementWidth, elementHeight) => {
	const viewport = getVisualViewport();
	const viewportWidth = viewport.width;
	const viewportHeight = viewport.height;
	const minX = viewport.offsetLeft + 16;
	const maxX = Math.max(minX, viewport.offsetLeft + viewportWidth - elementWidth - 16);
	const minY = viewport.offsetTop + 16;
	const maxY = Math.max(minY, viewport.offsetTop + viewportHeight - elementHeight - 16);
	if (edge === "top" || edge === "bottom") {
		const availableWidth = Math.max(0, viewportWidth - elementWidth - 16 * 2);
		return {
			x: Math.min(maxX, Math.max(minX, viewport.offsetLeft + 16 + availableWidth * ratio)),
			y: edge === "top" ? minY : maxY
		};
	}
	const availableHeight = Math.max(0, viewportHeight - elementHeight - 16 * 2);
	const positionY = Math.min(maxY, Math.max(minY, viewport.offsetTop + 16 + availableHeight * ratio));
	return {
		x: edge === "left" ? minX : maxX,
		y: positionY
	};
};
const getRatioFromPosition = (edge, positionX, positionY, elementWidth, elementHeight) => {
	const viewport = getVisualViewport();
	const viewportWidth = viewport.width;
	const viewportHeight = viewport.height;
	if (edge === "top" || edge === "bottom") {
		const availableWidth = viewportWidth - elementWidth - 16 * 2;
		if (availableWidth <= 0) return require_freeze_updates.TOOLBAR_DEFAULT_POSITION_RATIO;
		return Math.max(0, Math.min(1, (positionX - viewport.offsetLeft - 16) / availableWidth));
	}
	const availableHeight = viewportHeight - elementHeight - 16 * 2;
	if (availableHeight <= 0) return require_freeze_updates.TOOLBAR_DEFAULT_POSITION_RATIO;
	return Math.max(0, Math.min(1, (positionY - viewport.offsetTop - 16) / availableHeight));
};
const calculateExpandedPositionFromCollapsed = (collapsedPosition, edge, expandedDimensions, actualCollapsedWidth, actualCollapsedHeight) => {
	const viewport = getVisualViewport();
	const viewportWidth = viewport.width;
	const viewportHeight = viewport.height;
	const { width: expandedWidth, height: expandedHeight } = expandedDimensions;
	let newPosition;
	if (edge === "top" || edge === "bottom") {
		const xOffset = (expandedWidth - actualCollapsedWidth) / 2;
		newPosition = {
			x: clampToRange(collapsedPosition.x - xOffset, viewport.offsetLeft + 16, viewport.offsetLeft + viewportWidth - expandedWidth - 16),
			y: edge === "top" ? viewport.offsetTop + 16 : viewport.offsetTop + viewportHeight - expandedHeight - 16
		};
	} else {
		const yOffset = (expandedHeight - actualCollapsedHeight) / 2;
		const clampedY = clampToRange(collapsedPosition.y - yOffset, viewport.offsetTop + 16, viewport.offsetTop + viewportHeight - expandedHeight - 16);
		newPosition = {
			x: edge === "left" ? viewport.offsetLeft + 16 : viewport.offsetLeft + viewportWidth - expandedWidth - 16,
			y: clampedY
		};
	}
	const ratio = getRatioFromPosition(edge, newPosition.x, newPosition.y, expandedWidth, expandedHeight);
	return {
		position: newPosition,
		ratio
	};
};
const getCollapsedPosition = (edge, expandedPosition, expandedDimensions, collapsedDimensions) => {
	const viewport = getVisualViewport();
	const { width: expandedWidth, height: expandedHeight } = expandedDimensions;
	const { width: collapsedWidth, height: collapsedHeight } = collapsedDimensions;
	switch (edge) {
		case "top":
		case "bottom": {
			const xOffset = (expandedWidth - collapsedWidth) / 2;
			return {
				x: clampToRange(expandedPosition.x + xOffset, viewport.offsetLeft, viewport.offsetLeft + viewport.width - collapsedWidth),
				y: edge === "top" ? viewport.offsetTop : viewport.offsetTop + viewport.height - collapsedHeight
			};
		}
		case "left":
		case "right": {
			const yOffset = (expandedHeight - collapsedHeight) / 2;
			const clampedY = clampToRange(expandedPosition.y + yOffset, viewport.offsetTop, viewport.offsetTop + viewport.height - collapsedHeight);
			return {
				x: edge === "left" ? viewport.offsetLeft : viewport.offsetLeft + viewport.width - collapsedWidth,
				y: clampedY
			};
		}
	}
};
const getSnapPosition = (currentX, currentY, elementWidth, elementHeight, velocityX, velocityY) => {
	const viewport = getVisualViewport();
	const viewportWidth = viewport.width;
	const viewportHeight = viewport.height;
	const projectedX = currentX + velocityX * 150;
	const projectedY = currentY + velocityY * 150;
	const distanceToTop = projectedY - viewport.offsetTop + elementHeight / 2;
	const distanceToBottom = viewport.offsetTop + viewportHeight - projectedY - elementHeight / 2;
	const distanceToLeft = projectedX - viewport.offsetLeft + elementWidth / 2;
	const distanceToRight = viewport.offsetLeft + viewportWidth - projectedX - elementWidth / 2;
	const minDistance = Math.min(distanceToTop, distanceToBottom, distanceToLeft, distanceToRight);
	const clampX = (rawX) => clampToRange(rawX, viewport.offsetLeft + 16, viewport.offsetLeft + viewportWidth - elementWidth - 16);
	const clampY = (rawY) => clampToRange(rawY, viewport.offsetTop + 16, viewport.offsetTop + viewportHeight - elementHeight - 16);
	if (minDistance === distanceToTop) return {
		edge: "top",
		x: clampX(projectedX),
		y: viewport.offsetTop + 16
	};
	if (minDistance === distanceToLeft) return {
		edge: "left",
		x: viewport.offsetLeft + 16,
		y: clampY(projectedY)
	};
	if (minDistance === distanceToRight) return {
		edge: "right",
		x: viewport.offsetLeft + viewportWidth - elementWidth - 16,
		y: clampY(projectedY)
	};
	return {
		edge: "bottom",
		x: clampX(projectedX),
		y: viewport.offsetTop + viewportHeight - elementHeight - 16
	};
};

//#endregion
//#region src/utils/is-mac.ts
let cachedIsMac = null;
const getPlatformFromUserAgentData = () => {
	if (typeof navigator === "undefined") return null;
	if (!("userAgentData" in navigator)) return null;
	const userAgentData = navigator.userAgentData;
	if (typeof userAgentData !== "object" || userAgentData === null) return null;
	if (!("platform" in userAgentData)) return null;
	const platform = userAgentData.platform;
	if (typeof platform !== "string") return null;
	return platform;
};
const isMac = () => {
	if (cachedIsMac === null) {
		if (typeof navigator === "undefined") {
			cachedIsMac = false;
			return cachedIsMac;
		}
		const platform = navigator.platform ?? getPlatformFromUserAgentData() ?? navigator.userAgent;
		cachedIsMac = /Mac|iPhone|iPad|iPod/i.test(platform);
	}
	return cachedIsMac;
};

//#endregion
//#region src/utils/is-event-from-overlay.ts
const isEventFromOverlay = (event, attribute) => {
	try {
		return event.composedPath().some((target) => target instanceof HTMLElement && target.hasAttribute(attribute));
	} catch {
		return false;
	}
};

//#endregion
//#region src/utils/resolve-action-enabled.ts
const resolveBooleanEnabled = (enabled) => enabled ?? true;
const resolveActionEnabled = (action, context) => {
	if (typeof action.enabled === "function") {
		if (!context) return false;
		return action.enabled(context);
	}
	return resolveBooleanEnabled(action.enabled);
};

//#endregion
//#region src/components/toolbar/state.ts
const STORAGE_KEY = "react-grab-toolbar-state";
const loadToolbarState = () => {
	try {
		const serializedToolbarState = localStorage.getItem(STORAGE_KEY);
		if (!serializedToolbarState) return null;
		const parsed = JSON.parse(serializedToolbarState);
		if (typeof parsed !== "object" || parsed === null) return null;
		const record = parsed;
		return {
			edge: record.edge === "top" || record.edge === "bottom" || record.edge === "left" || record.edge === "right" ? record.edge : "bottom",
			ratio: typeof record.ratio === "number" ? record.ratio : require_freeze_updates.TOOLBAR_DEFAULT_POSITION_RATIO,
			collapsed: typeof record.collapsed === "boolean" ? record.collapsed : false,
			enabled: typeof record.enabled === "boolean" ? record.enabled : true,
			defaultAction: typeof record.defaultAction === "string" ? record.defaultAction : require_freeze_updates.DEFAULT_ACTION_ID
		};
	} catch (error) {
		console.warn("[react-grab] Failed to load toolbar state from localStorage:", error);
	}
	return null;
};
const saveToolbarState = (state) => {
	try {
		localStorage.setItem(STORAGE_KEY, JSON.stringify(state));
	} catch (error) {
		console.warn("[react-grab] Failed to save toolbar state to localStorage:", error);
	}
};

//#endregion
Object.defineProperty(exports, 'calculateExpandedPositionFromCollapsed', {
  enumerable: true,
  get: function () {
    return calculateExpandedPositionFromCollapsed;
  }
});
Object.defineProperty(exports, 'clampToRange', {
  enumerable: true,
  get: function () {
    return clampToRange;
  }
});
Object.defineProperty(exports, 'getCollapsedPosition', {
  enumerable: true,
  get: function () {
    return getCollapsedPosition;
  }
});
Object.defineProperty(exports, 'getPositionFromEdgeAndRatio', {
  enumerable: true,
  get: function () {
    return getPositionFromEdgeAndRatio;
  }
});
Object.defineProperty(exports, 'getRatioFromPosition', {
  enumerable: true,
  get: function () {
    return getRatioFromPosition;
  }
});
Object.defineProperty(exports, 'getSnapPosition', {
  enumerable: true,
  get: function () {
    return getSnapPosition;
  }
});
Object.defineProperty(exports, 'getVisualViewport', {
  enumerable: true,
  get: function () {
    return getVisualViewport;
  }
});
Object.defineProperty(exports, 'hasTextSelectionInInput', {
  enumerable: true,
  get: function () {
    return hasTextSelectionInInput;
  }
});
Object.defineProperty(exports, 'hasTextSelectionOnPage', {
  enumerable: true,
  get: function () {
    return hasTextSelectionOnPage;
  }
});
Object.defineProperty(exports, 'isElementConnected', {
  enumerable: true,
  get: function () {
    return isElementConnected;
  }
});
Object.defineProperty(exports, 'isEventFromOverlay', {
  enumerable: true,
  get: function () {
    return isEventFromOverlay;
  }
});
Object.defineProperty(exports, 'isKeyboardEventTriggeredByInput', {
  enumerable: true,
  get: function () {
    return isKeyboardEventTriggeredByInput;
  }
});
Object.defineProperty(exports, 'isMac', {
  enumerable: true,
  get: function () {
    return isMac;
  }
});
Object.defineProperty(exports, 'loadToolbarState', {
  enumerable: true,
  get: function () {
    return loadToolbarState;
  }
});
Object.defineProperty(exports, 'resolveActionEnabled', {
  enumerable: true,
  get: function () {
    return resolveActionEnabled;
  }
});
Object.defineProperty(exports, 'saveToolbarState', {
  enumerable: true,
  get: function () {
    return saveToolbarState;
  }
});
