// bippy@0.3.34/dist/index.cjs
/**
 * @license bippy
 *
 * Copyright (c) Aiden Bai
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const require_src = require('./src-J3uOnLH2.cjs');

exports.ActivityComponentTag = require_src.ActivityComponentTag;
exports.BIPPY_INSTRUMENTATION_STRING = require_src.BIPPY_INSTRUMENTATION_STRING;
exports.CONCURRENT_MODE_NUMBER = require_src.CONCURRENT_MODE_NUMBER;
exports.CONCURRENT_MODE_SYMBOL_STRING = require_src.CONCURRENT_MODE_SYMBOL_STRING;
exports.ClassComponentTag = require_src.ClassComponentTag;
exports.ContextConsumerTag = require_src.ContextConsumerTag;
exports.DEPRECATED_ASYNC_MODE_SYMBOL_STRING = require_src.DEPRECATED_ASYNC_MODE_SYMBOL_STRING;
exports.DehydratedSuspenseComponentTag = require_src.DehydratedSuspenseComponentTag;
exports.ELEMENT_TYPE_SYMBOL_STRING = require_src.ELEMENT_TYPE_SYMBOL_STRING;
exports.ForwardRefTag = require_src.ForwardRefTag;
exports.FragmentTag = require_src.FragmentTag;
exports.FunctionComponentTag = require_src.FunctionComponentTag;
exports.HostComponentTag = require_src.HostComponentTag;
exports.HostHoistableTag = require_src.HostHoistableTag;
exports.HostRootTag = require_src.HostRootTag;
exports.HostSingletonTag = require_src.HostSingletonTag;
exports.HostTextTag = require_src.HostTextTag;
exports.INSTALL_ERROR = require_src.INSTALL_ERROR;
exports.LazyComponentTag = require_src.LazyComponentTag;
exports.LegacyHiddenComponentTag = require_src.LegacyHiddenComponentTag;
exports.MemoComponentTag = require_src.MemoComponentTag;
exports.OffscreenComponentTag = require_src.OffscreenComponentTag;
exports.SimpleMemoComponentTag = require_src.SimpleMemoComponentTag;
exports.SuspenseComponentTag = require_src.SuspenseComponentTag;
exports.SuspenseListComponentTag = require_src.SuspenseListComponentTag;
exports.TRANSITIONAL_ELEMENT_TYPE_SYMBOL_STRING = require_src.TRANSITIONAL_ELEMENT_TYPE_SYMBOL_STRING;
exports.ViewTransitionComponentTag = require_src.ViewTransitionComponentTag;
exports._fiberRoots = require_src._fiberRoots;
exports._renderers = require_src._renderers;
exports.areFiberEqual = require_src.areFiberEqual;
exports.createFiberVisitor = require_src.createFiberVisitor;
exports.detectReactBuildType = require_src.detectReactBuildType;
exports.didFiberCommit = require_src.didFiberCommit;
exports.didFiberRender = require_src.didFiberRender;
exports.fiberIdMap = require_src.fiberIdMap;
exports.getDisplayName = require_src.getDisplayName;
exports.getFiberFromHostInstance = require_src.getFiberFromHostInstance;
exports.getFiberId = require_src.getFiberId;
exports.getFiberStack = require_src.getFiberStack;
exports.getLatestFiber = require_src.getLatestFiber;
exports.getMutatedHostFibers = require_src.getMutatedHostFibers;
exports.getNearestHostFiber = require_src.getNearestHostFiber;
exports.getNearestHostFibers = require_src.getNearestHostFibers;
exports.getRDTHook = require_src.getRDTHook;
exports.getTimings = require_src.getTimings;
exports.getType = require_src.getType;
exports.hasMemoCache = require_src.hasMemoCache;
exports.hasRDTHook = require_src.hasRDTHook;
exports.injectOverrideMethods = require_src.injectOverrideMethods;
exports.installRDTHook = require_src.installRDTHook;
exports.instrument = require_src.instrument;
exports.isClientEnvironment = require_src.isClientEnvironment;
exports.isCompositeFiber = require_src.isCompositeFiber;
exports.isHostFiber = require_src.isHostFiber;
exports.isInstrumentationActive = require_src.isInstrumentationActive;
exports.isReactRefresh = require_src.isReactRefresh;
exports.isRealReactDevtools = require_src.isRealReactDevtools;
exports.isValidElement = require_src.isValidElement;
exports.isValidFiber = require_src.isValidFiber;
exports.mountFiberRecursively = require_src.mountFiberRecursively;
exports.onCommitFiberRoot = require_src.onCommitFiberRoot;
exports.overrideContext = require_src.overrideContext;
exports.overrideHookState = require_src.overrideHookState;
exports.overrideProps = require_src.overrideProps;
exports.patchRDTHook = require_src.patchRDTHook;
exports.safelyInstallRDTHook = require_src.safelyInstallRDTHook;
exports.secure = require_src.secure;
exports.setFiberId = require_src.setFiberId;
exports.shouldFilterFiber = require_src.shouldFilterFiber;
exports.traverseContexts = require_src.traverseContexts;
exports.traverseFiber = require_src.traverseFiber;
exports.traverseFiberAsync = require_src.traverseFiberAsync;
exports.traverseFiberSync = require_src.traverseFiberSync;
exports.traverseProps = require_src.traverseProps;
exports.traverseRenderedFibers = require_src.traverseRenderedFibers;
exports.traverseState = require_src.traverseState;
exports.unmountFiber = require_src.unmountFiber;
exports.unmountFiberChildrenRecursively = require_src.unmountFiberChildrenRecursively;
exports.updateFiberRecursively = require_src.updateFiberRecursively;
exports.version = require_src.version;
