// bippy@0.3.34/dist/source.cjs
/**
 * @license bippy
 *
 * Copyright (c) Aiden Bai
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const require_source = require('./source-BAPwFAwe.cjs');
require('./src-J3uOnLH2.cjs');

exports.describeBuiltInComponentFrame = require_source.describeBuiltInComponentFrame;
exports.describeDebugInfoFrame = require_source.describeDebugInfoFrame;
exports.describeFiber = require_source.describeFiber;
exports.describeNativeComponentFrame = require_source.describeNativeComponentFrame;
exports.formatOwnerStack = require_source.formatOwnerStack;
exports.getCurrentDispatcher = require_source.getCurrentDispatcher;
exports.getFiberSource = require_source.getFiberSource;
exports.getFiberStackFrame = require_source.getFiberStackFrame;
exports.getFiberStackTrace = require_source.getFiberStackTrace;
exports.getOwnerStack = require_source.getOwnerStack;
exports.getSourceMap = require_source.getSourceMap;
exports.isCapitalized = require_source.isCapitalized;
exports.parseStackFrame = require_source.parseStackFrame;
exports.setCurrentDispatcher = require_source.setCurrentDispatcher;
